{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "833523af-4c7e-4a9e-91e5-b3733395407a",
   "metadata": {},
   "source": [
    "# Scpy, a single-cell gene differential expression analysis tool based on pytorch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a03a1080-508d-4fae-b089-a52d9bebce20",
   "metadata": {},
   "source": [
    "We use the data of Gasperini et al. [1] as an example to show how to use scpy for differential expression analysis, and compare it with the R package monocle3 [2,3].\n",
    "\n",
    "Gasperini et al. performed CRISPRi screening in the K562 cell line. In the at-scale experiment, they transfected 6,542 gRNA groups, and measured their effect on 13,135 highly expressed genes in 207,324 cells with single-cell RNA-seq.\n",
    "\n",
    "We downloaded the expression matrix, gRNA list and phenotype data of the at-scale experiment from GEO (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120861).\n",
    "\n",
    "First, we execute the R workflow."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc668f46-ed06-49e7-ad4a-12b7dcc795e9",
   "metadata": {},
   "source": [
    "## Monocle3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2a05694-f584-478b-90fa-20f7493a2b75",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘monocle3’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    exprs, fData, fData<-, pData, pData<-\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(monocle3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f7dbb73-1c97-4a34-a2fb-624d35d62346",
   "metadata": {},
   "outputs": [],
   "source": [
    "base <- '/home/jjpeng/data/yxlong/CRISPR/source/30849375/GEO/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c478b67-433a-4b42-ae93-a5872e39b536",
   "metadata": {},
   "source": [
    "### prepare featureData & phenoData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ccc650cb-0d80-4030-ada6-443dd768d483",
   "metadata": {},
   "outputs": [],
   "source": [
    "featureData <- read.table(gzfile(paste(base, 'GSE120861_at_scale_screen.genes.txt.gz', sep='')), row.names=1)\n",
    "featureData['id'] <- row.names(featureData)\n",
    "\n",
    "write.table(featureData, paste(base, 'at_scale.featureData.tsv', sep=''), sep='\\t')\n",
    "\n",
    "phenoData <- read.table(gzfile(paste(base, 'GSE120861_at_scale_screen.phenoData.txt.gz', sep='')), row.names=2)\n",
    "phenoData <- phenoData[c(1:4, 6:10, 14, 17)]\n",
    "colnames(phenoData) <- c('sample','total_umis','Size_Factor','gRNA_groups','barcodes','read_count','umi_count','proportion','guide_count','prep_batch','percent.mito')\n",
    "\n",
    "write.table(phenoData, paste(base, 'at_scale.phenoData.tsv', sep=''), sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ae04476-3f1a-415f-894d-ebc675bda783",
   "metadata": {},
   "source": [
    "### load CellDataSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "703853e3-6b9f-471d-834f-445e1fcf7bc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in new_cell_data_set(mat, cell_metadata = cell_annotations$metadata, :\n",
      "“Warning: gene_metadata must contain a column verbatim named 'gene_short_name' for certain functions.”\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: cell_data_set \n",
       "dim: 13135 207324 \n",
       "metadata(1): cds_version\n",
       "assays(1): counts\n",
       "rownames(13135): ENSG00000238009 ENSG00000237683 ... ENSG00000215615\n",
       "  ENSG00000215699\n",
       "rowData names(1): id\n",
       "colnames(207324): AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2\n",
       "  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2 ... TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9\n",
       "  TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9\n",
       "colData names(12): sample total_umis ... percent.mito n.umi\n",
       "reducedDimNames(0):\n",
       "mainExpName: NULL\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cds <- monocle3::load_mm_data(\n",
    "    mat_path = paste(base, 'GSE120861_at_scale_screen.exprs.mtx.gz', sep=''),\n",
    "    feature_anno_path = paste(base, 'at_scale.featureData.tsv', sep=''),\n",
    "    cell_anno_path = paste(base, 'at_scale.phenoData.tsv', sep=''),\n",
    "    header=TRUE\n",
    ")\n",
    "\n",
    "cds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "677a87f9-980f-4caa-abf4-e1fb2a2e2568",
   "metadata": {},
   "source": [
    "### choose bassik_mch, a non-targeting control gRNA group and 100 genes as an example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0909bf5c-ba3f-4327-a7af-004443a87d12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: cell_data_set \n",
       "dim: 100 207324 \n",
       "metadata(1): cds_version\n",
       "assays(1): counts\n",
       "rownames(100): ENSG00000238009 ENSG00000237683 ... ENSG00000116285\n",
       "  ENSG00000142599\n",
       "rowData names(1): id\n",
       "colnames(207324): AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2\n",
       "  AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2 ... TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9\n",
       "  TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9\n",
       "colData names(13): sample total_umis ... n.umi gRNA_group\n",
       "reducedDimNames(0):\n",
       "mainExpName: NULL\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes = fData(cds)[1:100, 'id']\n",
    "\n",
    "cds_subset <- cds[rowData(cds)$id %in% genes, ]\n",
    "\n",
    "pData(cds_subset)['gRNA_group'] <- as.integer(lapply(pData(cds)[,'gRNA_groups'], function(x) grepl('bassik_mch', x)))\n",
    "\n",
    "cds_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a32fab4f-71ba-4e5b-bbce-18402ce6b6af",
   "metadata": {},
   "source": [
    "### differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "06f19fdf-f70f-4c55-9189-0d4f51f6daba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time difference of 4.225742 mins\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 500 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene_id</th><th scope=col>status</th><th scope=col>term</th><th scope=col>estimate</th><th scope=col>std_err</th><th scope=col>test_val</th><th scope=col>p_value</th><th scope=col>normalized_effect</th><th scope=col>q_value</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSG00000238009</td><td>OK</td><td>(Intercept)           </td><td>-4.488624e+00</td><td>0.0588449363</td><td> -76.27885526</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000238009</td><td>OK</td><td>gRNA_group            </td><td> 9.159409e-02</td><td>0.3206558563</td><td>   0.28564609</td><td> 7.751492e-01</td><td> 7.142159e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000238009</td><td>OK</td><td>guide_count           </td><td>-4.850636e-04</td><td>0.0011927896</td><td>  -0.40666317</td><td> 6.842554e-01</td><td>-3.702237e-04</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000238009</td><td>OK</td><td>percent.mito          </td><td> 1.176492e+01</td><td>0.9931642069</td><td>  11.84589335</td><td> 2.259971e-32</td><td> 1.605482e+01</td><td> 8.361894e-31</td></tr>\n",
       "\t<tr><td>ENSG00000238009</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 3.846924e-02</td><td>0.0358320781</td><td>   1.07359778</td><td> 2.830030e-01</td><td> 2.963068e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000237683</td><td>OK</td><td>(Intercept)           </td><td>-3.598773e+00</td><td>0.0326945585</td><td>-110.07252209</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000237683</td><td>OK</td><td>gRNA_group            </td><td>-1.210731e-01</td><td>0.1899799610</td><td>  -0.63729420</td><td> 5.239332e-01</td><td>-1.258034e-01</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000237683</td><td>OK</td><td>guide_count           </td><td> 5.331953e-03</td><td>0.0006333395</td><td>   8.41879008</td><td> 3.803923e-17</td><td> 5.637260e-03</td><td> 2.929021e-15</td></tr>\n",
       "\t<tr><td>ENSG00000237683</td><td>OK</td><td>percent.mito          </td><td> 1.494807e+01</td><td>0.5411345368</td><td>  27.62357559</td><td>5.798296e-168</td><td> 2.111604e+01</td><td>3.305029e-166</td></tr>\n",
       "\t<tr><td>ENSG00000237683</td><td>OK</td><td>prep_batchprep_batch_2</td><td>-2.009985e-02</td><td>0.0195523402</td><td>  -1.02800240</td><td> 3.039487e-01</td><td>-2.117832e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000228463</td><td>OK</td><td>(Intercept)           </td><td>-5.546835e-01</td><td>0.0101808588</td><td> -54.48297722</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000228463</td><td>OK</td><td>gRNA_group            </td><td>-5.264484e-02</td><td>0.0619587382</td><td>  -0.84967572</td><td> 3.955054e-01</td><td>-7.461628e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000228463</td><td>OK</td><td>guide_count           </td><td>-1.015037e-03</td><td>0.0002140208</td><td>  -4.74270166</td><td> 2.108867e-06</td><td>-1.439312e-03</td><td> 1.328586e-04</td></tr>\n",
       "\t<tr><td>ENSG00000228463</td><td>OK</td><td>percent.mito          </td><td> 1.767073e+00</td><td>0.1800914084</td><td>   9.81208795</td><td> 9.988015e-23</td><td> 2.528726e+00</td><td> 3.296045e-21</td></tr>\n",
       "\t<tr><td>ENSG00000228463</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 1.078760e-02</td><td>0.0064311246</td><td>   1.67740494</td><td> 9.346332e-02</td><td> 1.529825e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000237094</td><td>OK</td><td>(Intercept)           </td><td>-3.933431e+00</td><td>0.0525286008</td><td> -74.88169384</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000237094</td><td>OK</td><td>gRNA_group            </td><td>-6.054079e-02</td><td>0.3206994621</td><td>  -0.18877733</td><td> 8.502673e-01</td><td>-5.721547e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000237094</td><td>OK</td><td>guide_count           </td><td>-3.168404e-03</td><td>0.0011039935</td><td>  -2.86994786</td><td> 4.105395e-03</td><td>-3.023916e-03</td><td> 1.847428e-01</td></tr>\n",
       "\t<tr><td>ENSG00000237094</td><td>OK</td><td>percent.mito          </td><td> 6.346012e+00</td><td>0.9138503874</td><td>   6.94425660</td><td> 3.804584e-12</td><td> 8.561320e+00</td><td> 9.131002e-11</td></tr>\n",
       "\t<tr><td>ENSG00000237094</td><td>OK</td><td>prep_batchprep_batch_2</td><td>-2.497970e-02</td><td>0.0328591292</td><td>  -0.76020585</td><td> 4.471315e-01</td><td>-2.375235e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000235373</td><td>OK</td><td>(Intercept)           </td><td>-3.369559e+00</td><td>0.0357204679</td><td> -94.33133704</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000235373</td><td>OK</td><td>gRNA_group            </td><td>-1.471128e-01</td><td>0.2188461941</td><td>  -0.67222013</td><td> 5.014436e-01</td><td>-1.616454e-01</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000235373</td><td>OK</td><td>guide_count           </td><td> 2.161045e-03</td><td>0.0007257931</td><td>   2.97749419</td><td> 2.906152e-03</td><td> 2.416201e-03</td><td> 1.365891e-01</td></tr>\n",
       "\t<tr><td>ENSG00000235373</td><td>OK</td><td>percent.mito          </td><td> 5.988558e+00</td><td>0.6159948836</td><td>   9.72176609</td><td> 2.435205e-22</td><td> 8.272608e+00</td><td> 7.792658e-21</td></tr>\n",
       "\t<tr><td>ENSG00000235373</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 9.034320e-02</td><td>0.0220711187</td><td>   4.09327686</td><td> 4.253194e-05</td><td> 1.019958e-01</td><td> 2.679512e-03</td></tr>\n",
       "\t<tr><td>ENSG00000228327</td><td>OK</td><td>(Intercept)           </td><td>-4.438321e+00</td><td>0.0725558586</td><td> -61.17109118</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000228327</td><td>OK</td><td>gRNA_group            </td><td> 3.016993e-01</td><td>0.3835027664</td><td>   0.78669399</td><td> 4.314610e-01</td><td> 2.518493e-01</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000228327</td><td>OK</td><td>guide_count           </td><td>-8.450155e-05</td><td>0.0015211272</td><td>  -0.05555193</td><td> 9.556988e-01</td><td>-6.602708e-05</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000228327</td><td>OK</td><td>percent.mito          </td><td>-3.973746e+00</td><td>1.2968734491</td><td>  -3.06409703</td><td> 2.183281e-03</td><td>-1.093669e+00</td><td> 3.056593e-02</td></tr>\n",
       "\t<tr><td>ENSG00000228327</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 3.816423e-01</td><td>0.0460905071</td><td>   8.28028097</td><td> 1.228850e-16</td><td> 3.238686e-01</td><td> 9.830799e-15</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>ENSG00000049246</td><td>OK</td><td>(Intercept)           </td><td>-1.785779e+00</td><td>1.539314e-02</td><td>-1.160113e+02</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000049246</td><td>OK</td><td>gRNA_group            </td><td>-7.872810e-04</td><td>8.860188e-02</td><td>-8.885602e-03</td><td> 9.929104e-01</td><td>-1.071854e-03</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000049246</td><td>OK</td><td>guide_count           </td><td> 4.231625e-06</td><td>3.127432e-04</td><td> 1.353067e-02</td><td> 9.892044e-01</td><td> 5.761327e-06</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000049246</td><td>OK</td><td>percent.mito          </td><td> 8.692280e+00</td><td>2.617372e-01</td><td> 3.320996e+01</td><td>7.732228e-242</td><td> 1.245675e+01</td><td>4.716659e-240</td></tr>\n",
       "\t<tr><td>ENSG00000049246</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 1.576104e-01</td><td>9.420459e-03</td><td> 1.673065e+01</td><td> 7.837326e-63</td><td> 2.154942e-01</td><td> 7.288713e-61</td></tr>\n",
       "\t<tr><td>ENSG00000236266</td><td>OK</td><td>(Intercept)           </td><td>-4.826811e+00</td><td>6.612007e-02</td><td>-7.300070e+01</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000236266</td><td>OK</td><td>gRNA_group            </td><td>-1.722361e-01</td><td>4.054157e-01</td><td>-4.248382e-01</td><td> 6.709547e-01</td><td>-1.052847e-01</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000236266</td><td>OK</td><td>guide_count           </td><td>-1.507530e-03</td><td>1.333675e-03</td><td>-1.130358e+00</td><td> 2.583255e-01</td><td>-9.670267e-04</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000236266</td><td>OK</td><td>percent.mito          </td><td> 1.523107e+01</td><td>1.101194e+00</td><td> 1.383142e+01</td><td> 1.647396e-43</td><td> 2.080507e+01</td><td> 6.589583e-42</td></tr>\n",
       "\t<tr><td>ENSG00000236266</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 6.460213e-03</td><td>3.986820e-02</td><td> 1.620393e-01</td><td> 8.712749e-01</td><td> 4.153168e-03</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000049249</td><td>OK</td><td>(Intercept)           </td><td>-1.885403e+00</td><td>4.228033e-02</td><td>-4.459291e+01</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000049249</td><td>OK</td><td>gRNA_group            </td><td>-5.375649e-01</td><td>3.066142e-01</td><td>-1.753229e+00</td><td> 7.956271e-02</td><td>-7.134159e-01</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000049249</td><td>OK</td><td>guide_count           </td><td>-1.453363e-03</td><td>9.273117e-04</td><td>-1.567286e+00</td><td> 1.170478e-01</td><td>-1.967056e-03</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000049249</td><td>OK</td><td>percent.mito          </td><td>-3.015517e+01</td><td>8.333400e-01</td><td>-3.618592e+01</td><td>1.013859e-286</td><td>-4.015853e+00</td><td>6.488695e-285</td></tr>\n",
       "\t<tr><td>ENSG00000049249</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 1.521366e-01</td><td>2.796150e-02</td><td> 5.440930e+00</td><td> 5.300309e-08</td><td> 2.068451e-01</td><td> 3.763220e-06</td></tr>\n",
       "\t<tr><td>ENSG00000116288</td><td>OK</td><td>(Intercept)           </td><td> 1.569867e+00</td><td>4.269091e-03</td><td> 3.677285e+02</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000116288</td><td>OK</td><td>gRNA_group            </td><td>-2.034713e-02</td><td>2.524741e-02</td><td>-8.059094e-01</td><td> 4.202951e-01</td><td>-2.929312e-02</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000116288</td><td>OK</td><td>guide_count           </td><td>-2.648644e-03</td><td>8.881516e-05</td><td>-2.982198e+01</td><td>2.027412e-195</td><td>-3.813240e-03</td><td>2.007138e-193</td></tr>\n",
       "\t<tr><td>ENSG00000116288</td><td>OK</td><td>percent.mito          </td><td> 1.024915e+01</td><td>7.446432e-02</td><td> 1.376385e+02</td><td> 0.000000e+00</td><td> 1.478341e+01</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000116288</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 1.048933e-01</td><td>2.661210e-03</td><td> 3.941563e+01</td><td> 0.000000e+00</td><td> 1.510307e-01</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000116285</td><td>OK</td><td>(Intercept)           </td><td>-2.207554e+00</td><td>3.548351e-02</td><td>-6.221353e+01</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000116285</td><td>OK</td><td>gRNA_group            </td><td>-5.187443e-01</td><td>2.583056e-01</td><td>-2.008258e+00</td><td> 4.461589e-02</td><td>-6.688591e-01</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000116285</td><td>OK</td><td>guide_count           </td><td>-1.793296e-03</td><td>7.692576e-04</td><td>-2.331203e+00</td><td> 1.974263e-02</td><td>-2.371348e-03</td><td> 8.291906e-01</td></tr>\n",
       "\t<tr><td>ENSG00000116285</td><td>OK</td><td>percent.mito          </td><td>-1.555488e+01</td><td>6.666009e-01</td><td>-2.333462e+01</td><td>1.974708e-120</td><td>-3.584590e+00</td><td>1.026848e-118</td></tr>\n",
       "\t<tr><td>ENSG00000116285</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 1.947013e-01</td><td>2.312359e-02</td><td> 8.420027e+00</td><td> 3.763985e-17</td><td> 2.594606e-01</td><td> 3.086468e-15</td></tr>\n",
       "\t<tr><td>ENSG00000142599</td><td>OK</td><td>(Intercept)           </td><td>-1.541760e+00</td><td>1.474325e-02</td><td>-1.045740e+02</td><td> 0.000000e+00</td><td> 0.000000e+00</td><td> 0.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000142599</td><td>OK</td><td>gRNA_group            </td><td>-1.125386e-01</td><td>9.025170e-02</td><td>-1.246941e+00</td><td> 2.124191e-01</td><td>-1.547075e-01</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000142599</td><td>OK</td><td>guide_count           </td><td>-3.210319e-04</td><td>3.039532e-04</td><td>-1.056189e+00</td><td> 2.908821e-01</td><td>-4.424720e-04</td><td> 1.000000e+00</td></tr>\n",
       "\t<tr><td>ENSG00000142599</td><td>OK</td><td>percent.mito          </td><td> 5.577083e+00</td><td>2.550158e-01</td><td> 2.186956e+01</td><td>5.064755e-106</td><td> 7.980398e+00</td><td>2.532378e-104</td></tr>\n",
       "\t<tr><td>ENSG00000142599</td><td>OK</td><td>prep_batchprep_batch_2</td><td> 1.095024e-01</td><td>9.143885e-03</td><td> 1.197548e+01</td><td> 4.776704e-33</td><td> 1.512830e-01</td><td> 4.155732e-31</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 500 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " gene\\_id & status & term & estimate & std\\_err & test\\_val & p\\_value & normalized\\_effect & q\\_value\\\\\n",
       " <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t ENSG00000238009 & OK & (Intercept)            & -4.488624e+00 & 0.0588449363 &  -76.27885526 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000238009 & OK & gRNA\\_group             &  9.159409e-02 & 0.3206558563 &    0.28564609 &  7.751492e-01 &  7.142159e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000238009 & OK & guide\\_count            & -4.850636e-04 & 0.0011927896 &   -0.40666317 &  6.842554e-01 & -3.702237e-04 &  1.000000e+00\\\\\n",
       "\t ENSG00000238009 & OK & percent.mito           &  1.176492e+01 & 0.9931642069 &   11.84589335 &  2.259971e-32 &  1.605482e+01 &  8.361894e-31\\\\\n",
       "\t ENSG00000238009 & OK & prep\\_batchprep\\_batch\\_2 &  3.846924e-02 & 0.0358320781 &    1.07359778 &  2.830030e-01 &  2.963068e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000237683 & OK & (Intercept)            & -3.598773e+00 & 0.0326945585 & -110.07252209 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000237683 & OK & gRNA\\_group             & -1.210731e-01 & 0.1899799610 &   -0.63729420 &  5.239332e-01 & -1.258034e-01 &  1.000000e+00\\\\\n",
       "\t ENSG00000237683 & OK & guide\\_count            &  5.331953e-03 & 0.0006333395 &    8.41879008 &  3.803923e-17 &  5.637260e-03 &  2.929021e-15\\\\\n",
       "\t ENSG00000237683 & OK & percent.mito           &  1.494807e+01 & 0.5411345368 &   27.62357559 & 5.798296e-168 &  2.111604e+01 & 3.305029e-166\\\\\n",
       "\t ENSG00000237683 & OK & prep\\_batchprep\\_batch\\_2 & -2.009985e-02 & 0.0195523402 &   -1.02800240 &  3.039487e-01 & -2.117832e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000228463 & OK & (Intercept)            & -5.546835e-01 & 0.0101808588 &  -54.48297722 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000228463 & OK & gRNA\\_group             & -5.264484e-02 & 0.0619587382 &   -0.84967572 &  3.955054e-01 & -7.461628e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000228463 & OK & guide\\_count            & -1.015037e-03 & 0.0002140208 &   -4.74270166 &  2.108867e-06 & -1.439312e-03 &  1.328586e-04\\\\\n",
       "\t ENSG00000228463 & OK & percent.mito           &  1.767073e+00 & 0.1800914084 &    9.81208795 &  9.988015e-23 &  2.528726e+00 &  3.296045e-21\\\\\n",
       "\t ENSG00000228463 & OK & prep\\_batchprep\\_batch\\_2 &  1.078760e-02 & 0.0064311246 &    1.67740494 &  9.346332e-02 &  1.529825e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000237094 & OK & (Intercept)            & -3.933431e+00 & 0.0525286008 &  -74.88169384 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000237094 & OK & gRNA\\_group             & -6.054079e-02 & 0.3206994621 &   -0.18877733 &  8.502673e-01 & -5.721547e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000237094 & OK & guide\\_count            & -3.168404e-03 & 0.0011039935 &   -2.86994786 &  4.105395e-03 & -3.023916e-03 &  1.847428e-01\\\\\n",
       "\t ENSG00000237094 & OK & percent.mito           &  6.346012e+00 & 0.9138503874 &    6.94425660 &  3.804584e-12 &  8.561320e+00 &  9.131002e-11\\\\\n",
       "\t ENSG00000237094 & OK & prep\\_batchprep\\_batch\\_2 & -2.497970e-02 & 0.0328591292 &   -0.76020585 &  4.471315e-01 & -2.375235e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000235373 & OK & (Intercept)            & -3.369559e+00 & 0.0357204679 &  -94.33133704 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000235373 & OK & gRNA\\_group             & -1.471128e-01 & 0.2188461941 &   -0.67222013 &  5.014436e-01 & -1.616454e-01 &  1.000000e+00\\\\\n",
       "\t ENSG00000235373 & OK & guide\\_count            &  2.161045e-03 & 0.0007257931 &    2.97749419 &  2.906152e-03 &  2.416201e-03 &  1.365891e-01\\\\\n",
       "\t ENSG00000235373 & OK & percent.mito           &  5.988558e+00 & 0.6159948836 &    9.72176609 &  2.435205e-22 &  8.272608e+00 &  7.792658e-21\\\\\n",
       "\t ENSG00000235373 & OK & prep\\_batchprep\\_batch\\_2 &  9.034320e-02 & 0.0220711187 &    4.09327686 &  4.253194e-05 &  1.019958e-01 &  2.679512e-03\\\\\n",
       "\t ENSG00000228327 & OK & (Intercept)            & -4.438321e+00 & 0.0725558586 &  -61.17109118 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000228327 & OK & gRNA\\_group             &  3.016993e-01 & 0.3835027664 &    0.78669399 &  4.314610e-01 &  2.518493e-01 &  1.000000e+00\\\\\n",
       "\t ENSG00000228327 & OK & guide\\_count            & -8.450155e-05 & 0.0015211272 &   -0.05555193 &  9.556988e-01 & -6.602708e-05 &  1.000000e+00\\\\\n",
       "\t ENSG00000228327 & OK & percent.mito           & -3.973746e+00 & 1.2968734491 &   -3.06409703 &  2.183281e-03 & -1.093669e+00 &  3.056593e-02\\\\\n",
       "\t ENSG00000228327 & OK & prep\\_batchprep\\_batch\\_2 &  3.816423e-01 & 0.0460905071 &    8.28028097 &  1.228850e-16 &  3.238686e-01 &  9.830799e-15\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t ENSG00000049246 & OK & (Intercept)            & -1.785779e+00 & 1.539314e-02 & -1.160113e+02 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000049246 & OK & gRNA\\_group             & -7.872810e-04 & 8.860188e-02 & -8.885602e-03 &  9.929104e-01 & -1.071854e-03 &  1.000000e+00\\\\\n",
       "\t ENSG00000049246 & OK & guide\\_count            &  4.231625e-06 & 3.127432e-04 &  1.353067e-02 &  9.892044e-01 &  5.761327e-06 &  1.000000e+00\\\\\n",
       "\t ENSG00000049246 & OK & percent.mito           &  8.692280e+00 & 2.617372e-01 &  3.320996e+01 & 7.732228e-242 &  1.245675e+01 & 4.716659e-240\\\\\n",
       "\t ENSG00000049246 & OK & prep\\_batchprep\\_batch\\_2 &  1.576104e-01 & 9.420459e-03 &  1.673065e+01 &  7.837326e-63 &  2.154942e-01 &  7.288713e-61\\\\\n",
       "\t ENSG00000236266 & OK & (Intercept)            & -4.826811e+00 & 6.612007e-02 & -7.300070e+01 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000236266 & OK & gRNA\\_group             & -1.722361e-01 & 4.054157e-01 & -4.248382e-01 &  6.709547e-01 & -1.052847e-01 &  1.000000e+00\\\\\n",
       "\t ENSG00000236266 & OK & guide\\_count            & -1.507530e-03 & 1.333675e-03 & -1.130358e+00 &  2.583255e-01 & -9.670267e-04 &  1.000000e+00\\\\\n",
       "\t ENSG00000236266 & OK & percent.mito           &  1.523107e+01 & 1.101194e+00 &  1.383142e+01 &  1.647396e-43 &  2.080507e+01 &  6.589583e-42\\\\\n",
       "\t ENSG00000236266 & OK & prep\\_batchprep\\_batch\\_2 &  6.460213e-03 & 3.986820e-02 &  1.620393e-01 &  8.712749e-01 &  4.153168e-03 &  1.000000e+00\\\\\n",
       "\t ENSG00000049249 & OK & (Intercept)            & -1.885403e+00 & 4.228033e-02 & -4.459291e+01 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000049249 & OK & gRNA\\_group             & -5.375649e-01 & 3.066142e-01 & -1.753229e+00 &  7.956271e-02 & -7.134159e-01 &  1.000000e+00\\\\\n",
       "\t ENSG00000049249 & OK & guide\\_count            & -1.453363e-03 & 9.273117e-04 & -1.567286e+00 &  1.170478e-01 & -1.967056e-03 &  1.000000e+00\\\\\n",
       "\t ENSG00000049249 & OK & percent.mito           & -3.015517e+01 & 8.333400e-01 & -3.618592e+01 & 1.013859e-286 & -4.015853e+00 & 6.488695e-285\\\\\n",
       "\t ENSG00000049249 & OK & prep\\_batchprep\\_batch\\_2 &  1.521366e-01 & 2.796150e-02 &  5.440930e+00 &  5.300309e-08 &  2.068451e-01 &  3.763220e-06\\\\\n",
       "\t ENSG00000116288 & OK & (Intercept)            &  1.569867e+00 & 4.269091e-03 &  3.677285e+02 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000116288 & OK & gRNA\\_group             & -2.034713e-02 & 2.524741e-02 & -8.059094e-01 &  4.202951e-01 & -2.929312e-02 &  1.000000e+00\\\\\n",
       "\t ENSG00000116288 & OK & guide\\_count            & -2.648644e-03 & 8.881516e-05 & -2.982198e+01 & 2.027412e-195 & -3.813240e-03 & 2.007138e-193\\\\\n",
       "\t ENSG00000116288 & OK & percent.mito           &  1.024915e+01 & 7.446432e-02 &  1.376385e+02 &  0.000000e+00 &  1.478341e+01 &  0.000000e+00\\\\\n",
       "\t ENSG00000116288 & OK & prep\\_batchprep\\_batch\\_2 &  1.048933e-01 & 2.661210e-03 &  3.941563e+01 &  0.000000e+00 &  1.510307e-01 &  0.000000e+00\\\\\n",
       "\t ENSG00000116285 & OK & (Intercept)            & -2.207554e+00 & 3.548351e-02 & -6.221353e+01 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000116285 & OK & gRNA\\_group             & -5.187443e-01 & 2.583056e-01 & -2.008258e+00 &  4.461589e-02 & -6.688591e-01 &  1.000000e+00\\\\\n",
       "\t ENSG00000116285 & OK & guide\\_count            & -1.793296e-03 & 7.692576e-04 & -2.331203e+00 &  1.974263e-02 & -2.371348e-03 &  8.291906e-01\\\\\n",
       "\t ENSG00000116285 & OK & percent.mito           & -1.555488e+01 & 6.666009e-01 & -2.333462e+01 & 1.974708e-120 & -3.584590e+00 & 1.026848e-118\\\\\n",
       "\t ENSG00000116285 & OK & prep\\_batchprep\\_batch\\_2 &  1.947013e-01 & 2.312359e-02 &  8.420027e+00 &  3.763985e-17 &  2.594606e-01 &  3.086468e-15\\\\\n",
       "\t ENSG00000142599 & OK & (Intercept)            & -1.541760e+00 & 1.474325e-02 & -1.045740e+02 &  0.000000e+00 &  0.000000e+00 &  0.000000e+00\\\\\n",
       "\t ENSG00000142599 & OK & gRNA\\_group             & -1.125386e-01 & 9.025170e-02 & -1.246941e+00 &  2.124191e-01 & -1.547075e-01 &  1.000000e+00\\\\\n",
       "\t ENSG00000142599 & OK & guide\\_count            & -3.210319e-04 & 3.039532e-04 & -1.056189e+00 &  2.908821e-01 & -4.424720e-04 &  1.000000e+00\\\\\n",
       "\t ENSG00000142599 & OK & percent.mito           &  5.577083e+00 & 2.550158e-01 &  2.186956e+01 & 5.064755e-106 &  7.980398e+00 & 2.532378e-104\\\\\n",
       "\t ENSG00000142599 & OK & prep\\_batchprep\\_batch\\_2 &  1.095024e-01 & 9.143885e-03 &  1.197548e+01 &  4.776704e-33 &  1.512830e-01 &  4.155732e-31\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 500 × 9\n",
       "\n",
       "| gene_id &lt;chr&gt; | status &lt;chr&gt; | term &lt;chr&gt; | estimate &lt;dbl&gt; | std_err &lt;dbl&gt; | test_val &lt;dbl&gt; | p_value &lt;dbl&gt; | normalized_effect &lt;dbl&gt; | q_value &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000238009 | OK | (Intercept)            | -4.488624e+00 | 0.0588449363 |  -76.27885526 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000238009 | OK | gRNA_group             |  9.159409e-02 | 0.3206558563 |    0.28564609 |  7.751492e-01 |  7.142159e-02 |  1.000000e+00 |\n",
       "| ENSG00000238009 | OK | guide_count            | -4.850636e-04 | 0.0011927896 |   -0.40666317 |  6.842554e-01 | -3.702237e-04 |  1.000000e+00 |\n",
       "| ENSG00000238009 | OK | percent.mito           |  1.176492e+01 | 0.9931642069 |   11.84589335 |  2.259971e-32 |  1.605482e+01 |  8.361894e-31 |\n",
       "| ENSG00000238009 | OK | prep_batchprep_batch_2 |  3.846924e-02 | 0.0358320781 |    1.07359778 |  2.830030e-01 |  2.963068e-02 |  1.000000e+00 |\n",
       "| ENSG00000237683 | OK | (Intercept)            | -3.598773e+00 | 0.0326945585 | -110.07252209 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000237683 | OK | gRNA_group             | -1.210731e-01 | 0.1899799610 |   -0.63729420 |  5.239332e-01 | -1.258034e-01 |  1.000000e+00 |\n",
       "| ENSG00000237683 | OK | guide_count            |  5.331953e-03 | 0.0006333395 |    8.41879008 |  3.803923e-17 |  5.637260e-03 |  2.929021e-15 |\n",
       "| ENSG00000237683 | OK | percent.mito           |  1.494807e+01 | 0.5411345368 |   27.62357559 | 5.798296e-168 |  2.111604e+01 | 3.305029e-166 |\n",
       "| ENSG00000237683 | OK | prep_batchprep_batch_2 | -2.009985e-02 | 0.0195523402 |   -1.02800240 |  3.039487e-01 | -2.117832e-02 |  1.000000e+00 |\n",
       "| ENSG00000228463 | OK | (Intercept)            | -5.546835e-01 | 0.0101808588 |  -54.48297722 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000228463 | OK | gRNA_group             | -5.264484e-02 | 0.0619587382 |   -0.84967572 |  3.955054e-01 | -7.461628e-02 |  1.000000e+00 |\n",
       "| ENSG00000228463 | OK | guide_count            | -1.015037e-03 | 0.0002140208 |   -4.74270166 |  2.108867e-06 | -1.439312e-03 |  1.328586e-04 |\n",
       "| ENSG00000228463 | OK | percent.mito           |  1.767073e+00 | 0.1800914084 |    9.81208795 |  9.988015e-23 |  2.528726e+00 |  3.296045e-21 |\n",
       "| ENSG00000228463 | OK | prep_batchprep_batch_2 |  1.078760e-02 | 0.0064311246 |    1.67740494 |  9.346332e-02 |  1.529825e-02 |  1.000000e+00 |\n",
       "| ENSG00000237094 | OK | (Intercept)            | -3.933431e+00 | 0.0525286008 |  -74.88169384 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000237094 | OK | gRNA_group             | -6.054079e-02 | 0.3206994621 |   -0.18877733 |  8.502673e-01 | -5.721547e-02 |  1.000000e+00 |\n",
       "| ENSG00000237094 | OK | guide_count            | -3.168404e-03 | 0.0011039935 |   -2.86994786 |  4.105395e-03 | -3.023916e-03 |  1.847428e-01 |\n",
       "| ENSG00000237094 | OK | percent.mito           |  6.346012e+00 | 0.9138503874 |    6.94425660 |  3.804584e-12 |  8.561320e+00 |  9.131002e-11 |\n",
       "| ENSG00000237094 | OK | prep_batchprep_batch_2 | -2.497970e-02 | 0.0328591292 |   -0.76020585 |  4.471315e-01 | -2.375235e-02 |  1.000000e+00 |\n",
       "| ENSG00000235373 | OK | (Intercept)            | -3.369559e+00 | 0.0357204679 |  -94.33133704 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000235373 | OK | gRNA_group             | -1.471128e-01 | 0.2188461941 |   -0.67222013 |  5.014436e-01 | -1.616454e-01 |  1.000000e+00 |\n",
       "| ENSG00000235373 | OK | guide_count            |  2.161045e-03 | 0.0007257931 |    2.97749419 |  2.906152e-03 |  2.416201e-03 |  1.365891e-01 |\n",
       "| ENSG00000235373 | OK | percent.mito           |  5.988558e+00 | 0.6159948836 |    9.72176609 |  2.435205e-22 |  8.272608e+00 |  7.792658e-21 |\n",
       "| ENSG00000235373 | OK | prep_batchprep_batch_2 |  9.034320e-02 | 0.0220711187 |    4.09327686 |  4.253194e-05 |  1.019958e-01 |  2.679512e-03 |\n",
       "| ENSG00000228327 | OK | (Intercept)            | -4.438321e+00 | 0.0725558586 |  -61.17109118 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000228327 | OK | gRNA_group             |  3.016993e-01 | 0.3835027664 |    0.78669399 |  4.314610e-01 |  2.518493e-01 |  1.000000e+00 |\n",
       "| ENSG00000228327 | OK | guide_count            | -8.450155e-05 | 0.0015211272 |   -0.05555193 |  9.556988e-01 | -6.602708e-05 |  1.000000e+00 |\n",
       "| ENSG00000228327 | OK | percent.mito           | -3.973746e+00 | 1.2968734491 |   -3.06409703 |  2.183281e-03 | -1.093669e+00 |  3.056593e-02 |\n",
       "| ENSG00000228327 | OK | prep_batchprep_batch_2 |  3.816423e-01 | 0.0460905071 |    8.28028097 |  1.228850e-16 |  3.238686e-01 |  9.830799e-15 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| ENSG00000049246 | OK | (Intercept)            | -1.785779e+00 | 1.539314e-02 | -1.160113e+02 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000049246 | OK | gRNA_group             | -7.872810e-04 | 8.860188e-02 | -8.885602e-03 |  9.929104e-01 | -1.071854e-03 |  1.000000e+00 |\n",
       "| ENSG00000049246 | OK | guide_count            |  4.231625e-06 | 3.127432e-04 |  1.353067e-02 |  9.892044e-01 |  5.761327e-06 |  1.000000e+00 |\n",
       "| ENSG00000049246 | OK | percent.mito           |  8.692280e+00 | 2.617372e-01 |  3.320996e+01 | 7.732228e-242 |  1.245675e+01 | 4.716659e-240 |\n",
       "| ENSG00000049246 | OK | prep_batchprep_batch_2 |  1.576104e-01 | 9.420459e-03 |  1.673065e+01 |  7.837326e-63 |  2.154942e-01 |  7.288713e-61 |\n",
       "| ENSG00000236266 | OK | (Intercept)            | -4.826811e+00 | 6.612007e-02 | -7.300070e+01 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000236266 | OK | gRNA_group             | -1.722361e-01 | 4.054157e-01 | -4.248382e-01 |  6.709547e-01 | -1.052847e-01 |  1.000000e+00 |\n",
       "| ENSG00000236266 | OK | guide_count            | -1.507530e-03 | 1.333675e-03 | -1.130358e+00 |  2.583255e-01 | -9.670267e-04 |  1.000000e+00 |\n",
       "| ENSG00000236266 | OK | percent.mito           |  1.523107e+01 | 1.101194e+00 |  1.383142e+01 |  1.647396e-43 |  2.080507e+01 |  6.589583e-42 |\n",
       "| ENSG00000236266 | OK | prep_batchprep_batch_2 |  6.460213e-03 | 3.986820e-02 |  1.620393e-01 |  8.712749e-01 |  4.153168e-03 |  1.000000e+00 |\n",
       "| ENSG00000049249 | OK | (Intercept)            | -1.885403e+00 | 4.228033e-02 | -4.459291e+01 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000049249 | OK | gRNA_group             | -5.375649e-01 | 3.066142e-01 | -1.753229e+00 |  7.956271e-02 | -7.134159e-01 |  1.000000e+00 |\n",
       "| ENSG00000049249 | OK | guide_count            | -1.453363e-03 | 9.273117e-04 | -1.567286e+00 |  1.170478e-01 | -1.967056e-03 |  1.000000e+00 |\n",
       "| ENSG00000049249 | OK | percent.mito           | -3.015517e+01 | 8.333400e-01 | -3.618592e+01 | 1.013859e-286 | -4.015853e+00 | 6.488695e-285 |\n",
       "| ENSG00000049249 | OK | prep_batchprep_batch_2 |  1.521366e-01 | 2.796150e-02 |  5.440930e+00 |  5.300309e-08 |  2.068451e-01 |  3.763220e-06 |\n",
       "| ENSG00000116288 | OK | (Intercept)            |  1.569867e+00 | 4.269091e-03 |  3.677285e+02 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000116288 | OK | gRNA_group             | -2.034713e-02 | 2.524741e-02 | -8.059094e-01 |  4.202951e-01 | -2.929312e-02 |  1.000000e+00 |\n",
       "| ENSG00000116288 | OK | guide_count            | -2.648644e-03 | 8.881516e-05 | -2.982198e+01 | 2.027412e-195 | -3.813240e-03 | 2.007138e-193 |\n",
       "| ENSG00000116288 | OK | percent.mito           |  1.024915e+01 | 7.446432e-02 |  1.376385e+02 |  0.000000e+00 |  1.478341e+01 |  0.000000e+00 |\n",
       "| ENSG00000116288 | OK | prep_batchprep_batch_2 |  1.048933e-01 | 2.661210e-03 |  3.941563e+01 |  0.000000e+00 |  1.510307e-01 |  0.000000e+00 |\n",
       "| ENSG00000116285 | OK | (Intercept)            | -2.207554e+00 | 3.548351e-02 | -6.221353e+01 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000116285 | OK | gRNA_group             | -5.187443e-01 | 2.583056e-01 | -2.008258e+00 |  4.461589e-02 | -6.688591e-01 |  1.000000e+00 |\n",
       "| ENSG00000116285 | OK | guide_count            | -1.793296e-03 | 7.692576e-04 | -2.331203e+00 |  1.974263e-02 | -2.371348e-03 |  8.291906e-01 |\n",
       "| ENSG00000116285 | OK | percent.mito           | -1.555488e+01 | 6.666009e-01 | -2.333462e+01 | 1.974708e-120 | -3.584590e+00 | 1.026848e-118 |\n",
       "| ENSG00000116285 | OK | prep_batchprep_batch_2 |  1.947013e-01 | 2.312359e-02 |  8.420027e+00 |  3.763985e-17 |  2.594606e-01 |  3.086468e-15 |\n",
       "| ENSG00000142599 | OK | (Intercept)            | -1.541760e+00 | 1.474325e-02 | -1.045740e+02 |  0.000000e+00 |  0.000000e+00 |  0.000000e+00 |\n",
       "| ENSG00000142599 | OK | gRNA_group             | -1.125386e-01 | 9.025170e-02 | -1.246941e+00 |  2.124191e-01 | -1.547075e-01 |  1.000000e+00 |\n",
       "| ENSG00000142599 | OK | guide_count            | -3.210319e-04 | 3.039532e-04 | -1.056189e+00 |  2.908821e-01 | -4.424720e-04 |  1.000000e+00 |\n",
       "| ENSG00000142599 | OK | percent.mito           |  5.577083e+00 | 2.550158e-01 |  2.186956e+01 | 5.064755e-106 |  7.980398e+00 | 2.532378e-104 |\n",
       "| ENSG00000142599 | OK | prep_batchprep_batch_2 |  1.095024e-01 | 9.143885e-03 |  1.197548e+01 |  4.776704e-33 |  1.512830e-01 |  4.155732e-31 |\n",
       "\n"
      ],
      "text/plain": [
       "    gene_id         status term                   estimate      std_err     \n",
       "1   ENSG00000238009 OK     (Intercept)            -4.488624e+00 0.0588449363\n",
       "2   ENSG00000238009 OK     gRNA_group              9.159409e-02 0.3206558563\n",
       "3   ENSG00000238009 OK     guide_count            -4.850636e-04 0.0011927896\n",
       "4   ENSG00000238009 OK     percent.mito            1.176492e+01 0.9931642069\n",
       "5   ENSG00000238009 OK     prep_batchprep_batch_2  3.846924e-02 0.0358320781\n",
       "6   ENSG00000237683 OK     (Intercept)            -3.598773e+00 0.0326945585\n",
       "7   ENSG00000237683 OK     gRNA_group             -1.210731e-01 0.1899799610\n",
       "8   ENSG00000237683 OK     guide_count             5.331953e-03 0.0006333395\n",
       "9   ENSG00000237683 OK     percent.mito            1.494807e+01 0.5411345368\n",
       "10  ENSG00000237683 OK     prep_batchprep_batch_2 -2.009985e-02 0.0195523402\n",
       "11  ENSG00000228463 OK     (Intercept)            -5.546835e-01 0.0101808588\n",
       "12  ENSG00000228463 OK     gRNA_group             -5.264484e-02 0.0619587382\n",
       "13  ENSG00000228463 OK     guide_count            -1.015037e-03 0.0002140208\n",
       "14  ENSG00000228463 OK     percent.mito            1.767073e+00 0.1800914084\n",
       "15  ENSG00000228463 OK     prep_batchprep_batch_2  1.078760e-02 0.0064311246\n",
       "16  ENSG00000237094 OK     (Intercept)            -3.933431e+00 0.0525286008\n",
       "17  ENSG00000237094 OK     gRNA_group             -6.054079e-02 0.3206994621\n",
       "18  ENSG00000237094 OK     guide_count            -3.168404e-03 0.0011039935\n",
       "19  ENSG00000237094 OK     percent.mito            6.346012e+00 0.9138503874\n",
       "20  ENSG00000237094 OK     prep_batchprep_batch_2 -2.497970e-02 0.0328591292\n",
       "21  ENSG00000235373 OK     (Intercept)            -3.369559e+00 0.0357204679\n",
       "22  ENSG00000235373 OK     gRNA_group             -1.471128e-01 0.2188461941\n",
       "23  ENSG00000235373 OK     guide_count             2.161045e-03 0.0007257931\n",
       "24  ENSG00000235373 OK     percent.mito            5.988558e+00 0.6159948836\n",
       "25  ENSG00000235373 OK     prep_batchprep_batch_2  9.034320e-02 0.0220711187\n",
       "26  ENSG00000228327 OK     (Intercept)            -4.438321e+00 0.0725558586\n",
       "27  ENSG00000228327 OK     gRNA_group              3.016993e-01 0.3835027664\n",
       "28  ENSG00000228327 OK     guide_count            -8.450155e-05 0.0015211272\n",
       "29  ENSG00000228327 OK     percent.mito           -3.973746e+00 1.2968734491\n",
       "30  ENSG00000228327 OK     prep_batchprep_batch_2  3.816423e-01 0.0460905071\n",
       "⋮   ⋮               ⋮      ⋮                      ⋮             ⋮           \n",
       "471 ENSG00000049246 OK     (Intercept)            -1.785779e+00 1.539314e-02\n",
       "472 ENSG00000049246 OK     gRNA_group             -7.872810e-04 8.860188e-02\n",
       "473 ENSG00000049246 OK     guide_count             4.231625e-06 3.127432e-04\n",
       "474 ENSG00000049246 OK     percent.mito            8.692280e+00 2.617372e-01\n",
       "475 ENSG00000049246 OK     prep_batchprep_batch_2  1.576104e-01 9.420459e-03\n",
       "476 ENSG00000236266 OK     (Intercept)            -4.826811e+00 6.612007e-02\n",
       "477 ENSG00000236266 OK     gRNA_group             -1.722361e-01 4.054157e-01\n",
       "478 ENSG00000236266 OK     guide_count            -1.507530e-03 1.333675e-03\n",
       "479 ENSG00000236266 OK     percent.mito            1.523107e+01 1.101194e+00\n",
       "480 ENSG00000236266 OK     prep_batchprep_batch_2  6.460213e-03 3.986820e-02\n",
       "481 ENSG00000049249 OK     (Intercept)            -1.885403e+00 4.228033e-02\n",
       "482 ENSG00000049249 OK     gRNA_group             -5.375649e-01 3.066142e-01\n",
       "483 ENSG00000049249 OK     guide_count            -1.453363e-03 9.273117e-04\n",
       "484 ENSG00000049249 OK     percent.mito           -3.015517e+01 8.333400e-01\n",
       "485 ENSG00000049249 OK     prep_batchprep_batch_2  1.521366e-01 2.796150e-02\n",
       "486 ENSG00000116288 OK     (Intercept)             1.569867e+00 4.269091e-03\n",
       "487 ENSG00000116288 OK     gRNA_group             -2.034713e-02 2.524741e-02\n",
       "488 ENSG00000116288 OK     guide_count            -2.648644e-03 8.881516e-05\n",
       "489 ENSG00000116288 OK     percent.mito            1.024915e+01 7.446432e-02\n",
       "490 ENSG00000116288 OK     prep_batchprep_batch_2  1.048933e-01 2.661210e-03\n",
       "491 ENSG00000116285 OK     (Intercept)            -2.207554e+00 3.548351e-02\n",
       "492 ENSG00000116285 OK     gRNA_group             -5.187443e-01 2.583056e-01\n",
       "493 ENSG00000116285 OK     guide_count            -1.793296e-03 7.692576e-04\n",
       "494 ENSG00000116285 OK     percent.mito           -1.555488e+01 6.666009e-01\n",
       "495 ENSG00000116285 OK     prep_batchprep_batch_2  1.947013e-01 2.312359e-02\n",
       "496 ENSG00000142599 OK     (Intercept)            -1.541760e+00 1.474325e-02\n",
       "497 ENSG00000142599 OK     gRNA_group             -1.125386e-01 9.025170e-02\n",
       "498 ENSG00000142599 OK     guide_count            -3.210319e-04 3.039532e-04\n",
       "499 ENSG00000142599 OK     percent.mito            5.577083e+00 2.550158e-01\n",
       "500 ENSG00000142599 OK     prep_batchprep_batch_2  1.095024e-01 9.143885e-03\n",
       "    test_val      p_value       normalized_effect q_value      \n",
       "1    -76.27885526  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "2      0.28564609  7.751492e-01  7.142159e-02      1.000000e+00\n",
       "3     -0.40666317  6.842554e-01 -3.702237e-04      1.000000e+00\n",
       "4     11.84589335  2.259971e-32  1.605482e+01      8.361894e-31\n",
       "5      1.07359778  2.830030e-01  2.963068e-02      1.000000e+00\n",
       "6   -110.07252209  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "7     -0.63729420  5.239332e-01 -1.258034e-01      1.000000e+00\n",
       "8      8.41879008  3.803923e-17  5.637260e-03      2.929021e-15\n",
       "9     27.62357559 5.798296e-168  2.111604e+01     3.305029e-166\n",
       "10    -1.02800240  3.039487e-01 -2.117832e-02      1.000000e+00\n",
       "11   -54.48297722  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "12    -0.84967572  3.955054e-01 -7.461628e-02      1.000000e+00\n",
       "13    -4.74270166  2.108867e-06 -1.439312e-03      1.328586e-04\n",
       "14     9.81208795  9.988015e-23  2.528726e+00      3.296045e-21\n",
       "15     1.67740494  9.346332e-02  1.529825e-02      1.000000e+00\n",
       "16   -74.88169384  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "17    -0.18877733  8.502673e-01 -5.721547e-02      1.000000e+00\n",
       "18    -2.86994786  4.105395e-03 -3.023916e-03      1.847428e-01\n",
       "19     6.94425660  3.804584e-12  8.561320e+00      9.131002e-11\n",
       "20    -0.76020585  4.471315e-01 -2.375235e-02      1.000000e+00\n",
       "21   -94.33133704  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "22    -0.67222013  5.014436e-01 -1.616454e-01      1.000000e+00\n",
       "23     2.97749419  2.906152e-03  2.416201e-03      1.365891e-01\n",
       "24     9.72176609  2.435205e-22  8.272608e+00      7.792658e-21\n",
       "25     4.09327686  4.253194e-05  1.019958e-01      2.679512e-03\n",
       "26   -61.17109118  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "27     0.78669399  4.314610e-01  2.518493e-01      1.000000e+00\n",
       "28    -0.05555193  9.556988e-01 -6.602708e-05      1.000000e+00\n",
       "29    -3.06409703  2.183281e-03 -1.093669e+00      3.056593e-02\n",
       "30     8.28028097  1.228850e-16  3.238686e-01      9.830799e-15\n",
       "⋮   ⋮             ⋮             ⋮                 ⋮            \n",
       "471 -1.160113e+02  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "472 -8.885602e-03  9.929104e-01 -1.071854e-03      1.000000e+00\n",
       "473  1.353067e-02  9.892044e-01  5.761327e-06      1.000000e+00\n",
       "474  3.320996e+01 7.732228e-242  1.245675e+01     4.716659e-240\n",
       "475  1.673065e+01  7.837326e-63  2.154942e-01      7.288713e-61\n",
       "476 -7.300070e+01  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "477 -4.248382e-01  6.709547e-01 -1.052847e-01      1.000000e+00\n",
       "478 -1.130358e+00  2.583255e-01 -9.670267e-04      1.000000e+00\n",
       "479  1.383142e+01  1.647396e-43  2.080507e+01      6.589583e-42\n",
       "480  1.620393e-01  8.712749e-01  4.153168e-03      1.000000e+00\n",
       "481 -4.459291e+01  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "482 -1.753229e+00  7.956271e-02 -7.134159e-01      1.000000e+00\n",
       "483 -1.567286e+00  1.170478e-01 -1.967056e-03      1.000000e+00\n",
       "484 -3.618592e+01 1.013859e-286 -4.015853e+00     6.488695e-285\n",
       "485  5.440930e+00  5.300309e-08  2.068451e-01      3.763220e-06\n",
       "486  3.677285e+02  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "487 -8.059094e-01  4.202951e-01 -2.929312e-02      1.000000e+00\n",
       "488 -2.982198e+01 2.027412e-195 -3.813240e-03     2.007138e-193\n",
       "489  1.376385e+02  0.000000e+00  1.478341e+01      0.000000e+00\n",
       "490  3.941563e+01  0.000000e+00  1.510307e-01      0.000000e+00\n",
       "491 -6.221353e+01  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "492 -2.008258e+00  4.461589e-02 -6.688591e-01      1.000000e+00\n",
       "493 -2.331203e+00  1.974263e-02 -2.371348e-03      8.291906e-01\n",
       "494 -2.333462e+01 1.974708e-120 -3.584590e+00     1.026848e-118\n",
       "495  8.420027e+00  3.763985e-17  2.594606e-01      3.086468e-15\n",
       "496 -1.045740e+02  0.000000e+00  0.000000e+00      0.000000e+00\n",
       "497 -1.246941e+00  2.124191e-01 -1.547075e-01      1.000000e+00\n",
       "498 -1.056189e+00  2.908821e-01 -4.424720e-04      1.000000e+00\n",
       "499  2.186956e+01 5.064755e-106  7.980398e+00     2.532378e-104\n",
       "500  1.197548e+01  4.776704e-33  1.512830e-01      4.155732e-31"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "formula <- \"~gRNA_group+guide_count+percent.mito+prep_batch\"\n",
    "\n",
    "tick <- Sys.time()\n",
    "fits <- monocle3::fit_models(cds_subset, expression_family = 'negbinomial', model_formula_str = formula)\n",
    "print(Sys.time() - tick)\n",
    "\n",
    "coef <- coefficient_table(fits)[c('gene_id','status','term','estimate','std_err','test_val','p_value','normalized_effect','q_value')]\n",
    "\n",
    "write.csv(coef, 'bassik_mch-100.csv', row.names=FALSE)\n",
    "\n",
    "coef"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c77b6ee-5749-4a85-a403-dc70e057084c",
   "metadata": {},
   "source": [
    "It takes 2.53 seconds to test a relationship pair between a set of variables and a gene, on average.\n",
    "\n",
    "This means that it will take **6.9 years** to complete the test of 85,929,170 pairs of interactions between 6,542 gRNA groups and 13,135 genes ! ! !\n",
    "\n",
    "In the paper, Gasperini et al. measured the relationship between each gRNA group and surrounding 1 Mbp genes, a total of 629,515 pairs. They also used negbinomial.size family to increase speed by reducing accuracy.\n",
    "\n",
    "Now, let's move to python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba4cba6b-5c99-43d5-83c7-7069b03f69c9",
   "metadata": {},
   "source": [
    "## Statsmodels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da38a65e-36d3-4fa5-a06c-1178de4a10c4",
   "metadata": {},
   "source": [
    "Scpy is deeply inspired by statsmodels [4] and relies on it to solve optimizations.\n",
    "\n",
    "We first show how to use statsmodels for differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8621a41f-a193-403f-858a-754441de3809",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "from tqdm import tqdm\n",
    "import time\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "import statsmodels as sm\n",
    "import statsmodels.api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b36fb196-8c0f-4d01-b242-3000d6169a0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "base = '/home/jjpeng/data/yxlong/CRISPR/source/30849375/GEO/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b205f35-50f1-4377-94a5-1a14906e37dd",
   "metadata": {},
   "source": [
    "### load featureData & phenoData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "adc9dcfa-831c-4441-91c8-87bb377057cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "grnas = pd.read_csv(base + 'GSE120861_grna_groups.at_scale.txt.gz', sep = '\\t', header=None)[1]\n",
    "cells = pd.read_csv(base + 'GSE120861_at_scale_screen.cells.txt.gz', sep = '\\t', header=None)[0]\n",
    "genes = pd.read_csv(base + 'GSE120861_at_scale_screen.genes.txt.gz', sep = '\\t', header=None)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8da83f4c-6943-4805-81ff-48daae2a062a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>total_umis</th>\n",
       "      <th>gRNA_groups</th>\n",
       "      <th>read_count</th>\n",
       "      <th>umi_count</th>\n",
       "      <th>proportion</th>\n",
       "      <th>guide_count</th>\n",
       "      <th>prep_batch</th>\n",
       "      <th>percent.mito</th>\n",
       "      <th>size_factor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>17572.0</td>\n",
       "      <td>chr10.845_top_two_chr1.11183_top_two_chr1.1129...</td>\n",
       "      <td>14135.0</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.969819</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.058787</td>\n",
       "      <td>1.084235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>8923.0</td>\n",
       "      <td>chr1.12695_top_two_chr11.3294_top_two_chr1.679...</td>\n",
       "      <td>4329.0</td>\n",
       "      <td>293.0</td>\n",
       "      <td>0.844380</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.036087</td>\n",
       "      <td>0.550571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>14637.0</td>\n",
       "      <td>ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....</td>\n",
       "      <td>12362.0</td>\n",
       "      <td>884.0</td>\n",
       "      <td>0.950538</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.069823</td>\n",
       "      <td>0.903138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>22798.0</td>\n",
       "      <td>C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...</td>\n",
       "      <td>7459.0</td>\n",
       "      <td>544.0</td>\n",
       "      <td>0.939551</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.026187</td>\n",
       "      <td>1.406692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>1A_1_SI-GA-E2</td>\n",
       "      <td>10136.0</td>\n",
       "      <td>chr10.185_top_two_chr10.484_top_two_chr11.4167...</td>\n",
       "      <td>14831.0</td>\n",
       "      <td>1054.0</td>\n",
       "      <td>0.959927</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.007991</td>\n",
       "      <td>0.625416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>17938.0</td>\n",
       "      <td>BRIX1_TSS_chr10.2059_top_two_chr10.350_top_two...</td>\n",
       "      <td>4570.0</td>\n",
       "      <td>394.0</td>\n",
       "      <td>0.856522</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.060598</td>\n",
       "      <td>1.106818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>16543.0</td>\n",
       "      <td>chr10.221_top_two_chr11.3853_top_two_chr11.498...</td>\n",
       "      <td>4311.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>0.861915</td>\n",
       "      <td>33.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.057789</td>\n",
       "      <td>1.020743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>15009.0</td>\n",
       "      <td>chr12.3400_top_two_chr4.1281_second_two_ID3_TSS</td>\n",
       "      <td>669.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.876923</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.049037</td>\n",
       "      <td>0.926092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>5149.0</td>\n",
       "      <td>chr7.1014_top_two</td>\n",
       "      <td>77.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.086420</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000971</td>\n",
       "      <td>0.317706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</th>\n",
       "      <td>2B_8_SI-GA-H9</td>\n",
       "      <td>9503.0</td>\n",
       "      <td>chr11.2790_top_two_chr12.306_top_two_chr12.474...</td>\n",
       "      <td>4630.0</td>\n",
       "      <td>410.0</td>\n",
       "      <td>0.813492</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.028623</td>\n",
       "      <td>0.586358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>207324 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         sample  total_umis  \\\n",
       "cell                                                          \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  1A_1_SI-GA-E2     17572.0   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  1A_1_SI-GA-E2      8923.0   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  1A_1_SI-GA-E2     14637.0   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  1A_1_SI-GA-E2     22798.0   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  1A_1_SI-GA-E2     10136.0   \n",
       "...                                         ...         ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9  2B_8_SI-GA-H9     17938.0   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9  2B_8_SI-GA-H9     16543.0   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9  2B_8_SI-GA-H9     15009.0   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9  2B_8_SI-GA-H9      5149.0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9  2B_8_SI-GA-H9      9503.0   \n",
       "\n",
       "                                                                        gRNA_groups  \\\n",
       "cell                                                                                  \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2  chr10.845_top_two_chr1.11183_top_two_chr1.1129...   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2  chr1.12695_top_two_chr11.3294_top_two_chr1.679...   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2  ALDH1A2_TSS_BRI3_TSS_chr10.1918_top_two_chr10....   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2  C16orf91_TSS_chr1.11332_top_two_chr1.1933_top_...   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2  chr10.185_top_two_chr10.484_top_two_chr11.4167...   \n",
       "...                                                                             ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9  BRIX1_TSS_chr10.2059_top_two_chr10.350_top_two...   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9  chr10.221_top_two_chr11.3853_top_two_chr11.498...   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9    chr12.3400_top_two_chr4.1281_second_two_ID3_TSS   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                                  chr7.1014_top_two   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9  chr11.2790_top_two_chr12.306_top_two_chr12.474...   \n",
       "\n",
       "                                  read_count  umi_count  proportion  \\\n",
       "cell                                                                  \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2     14135.0      964.0    0.969819   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2      4329.0      293.0    0.844380   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2     12362.0      884.0    0.950538   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2      7459.0      544.0    0.939551   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2     14831.0     1054.0    0.959927   \n",
       "...                                      ...        ...         ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9      4570.0      394.0    0.856522   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9      4311.0      387.0    0.861915   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9       669.0       57.0    0.876923   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9        77.0        7.0    0.086420   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9      4630.0      410.0    0.813492   \n",
       "\n",
       "                                  guide_count  prep_batch  percent.mito  \\\n",
       "cell                                                                      \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2         67.0           0      0.058787   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2         26.0           0      0.036087   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2         61.0           0      0.069823   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2         39.0           0      0.026187   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2         37.0           0      0.007991   \n",
       "...                                       ...         ...           ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9         31.0           1      0.060598   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9         33.0           1      0.057789   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9          3.0           1      0.049037   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9          1.0           1      0.000971   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9         35.0           1      0.028623   \n",
       "\n",
       "                                  size_factor  \n",
       "cell                                           \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2     1.084235  \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2     0.550571  \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2     0.903138  \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2     1.406692  \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2     0.625416  \n",
       "...                                       ...  \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9     1.106818  \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9     1.020743  \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9     0.926092  \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9     0.317706  \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9     0.586358  \n",
       "\n",
       "[207324 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenoData = pd.read_csv(base + 'GSE120861_at_scale_screen.phenoData.txt.gz', sep=' ', header=None, usecols=[0,1,2,5,7,8,9,10,14,17])\n",
    "phenoData.columns = ['sample','cell','total_umis','gRNA_groups','read_count','umi_count','proportion','guide_count','prep_batch','percent.mito']\n",
    "phenoData.set_index('cell', inplace=True)\n",
    "\n",
    "phenoData['guide_count'].fillna(0, inplace=True)\n",
    "phenoData['size_factor'] = phenoData['total_umis'] / np.exp(np.mean(np.log(phenoData['total_umis'])))\n",
    "phenoData['prep_batch'] = phenoData.eval('prep_batch != \"prep_batch_1\"').astype(int)\n",
    "\n",
    "phenoData"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "466c578c-1f82-4bbb-a823-2d38b0844199",
   "metadata": {},
   "source": [
    "### prepare & load expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4b6e386-acc6-49b4-904c-909cd228a0a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(base + 'at_scale.exprs.npy'):\n",
    "\n",
    "    mtx = pd.read_csv(base + 'GSE120861_at_scale_screen.exprs.mtx.gz', sep = ' ', header=None, skiprows=2, compression=None)\n",
    "    coo = sp.sparse.coo_matrix((mtx[2], (mtx[0] - 1, mtx[1] - 1)), shape=(genes.shape[0], cells.shape[0]), dtype=np.int16)\n",
    "    np.save(base + 'at_scale.exprs.npy', coo)\n",
    "    \n",
    "else:\n",
    "\n",
    "    coo = np.load(base + 'at_scale.exprs.npy', allow_pickle=True).item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e795a20-903f-4868-9a5e-be039a84e171",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <th>ENSG00000237683</th>\n",
       "      <th>ENSG00000228463</th>\n",
       "      <th>ENSG00000237094</th>\n",
       "      <th>ENSG00000235373</th>\n",
       "      <th>ENSG00000228327</th>\n",
       "      <th>ENSG00000237491</th>\n",
       "      <th>ENSG00000225880</th>\n",
       "      <th>ENSG00000230368</th>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <th>ENSG00000215750</th>\n",
       "      <th>ENSG00000215689</th>\n",
       "      <th>ENSG00000215781</th>\n",
       "      <th>ENSG00000220023</th>\n",
       "      <th>ENSG00000215615</th>\n",
       "      <th>ENSG00000215699</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>115</td>\n",
       "      <td>20</td>\n",
       "      <td>5</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>45</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAACAACA-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>156</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>119</td>\n",
       "      <td>30</td>\n",
       "      <td>6</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>113</td>\n",
       "      <td>32</td>\n",
       "      <td>13</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>93</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>207324 rows × 13135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                 ENSG00000238009  ENSG00000237683  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                0                1   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0                0   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                0                1   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0                0   \n",
       "\n",
       "0                                 ENSG00000228463  ENSG00000237094  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                1                0   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                1                0   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                1                1   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                1                0   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                2                0   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                2                0   \n",
       "\n",
       "0                                 ENSG00000235373  ENSG00000228327  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                1                0   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0                0   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0                0   \n",
       "\n",
       "0                                 ENSG00000237491  ENSG00000225880  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0                0   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                1                0   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0                0   \n",
       "\n",
       "0                                 ENSG00000230368  ENSG00000188976  ...  \\\n",
       "0                                                                   ...   \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                0                0  ...   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                0  ...   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                0                3  ...   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                0                1  ...   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0                0  ...   \n",
       "...                                           ...              ...  ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                0                1  ...   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                0                3  ...   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                0                4  ...   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                1  ...   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0                2  ...   \n",
       "\n",
       "0                                 ENSG00000198886  ENSG00000198786  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2              115               20   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2               45               17   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2              156               20   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2               66               24   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                8                4   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9              119               30   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9              113               32   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9               93               21   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                1                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9               33               13   \n",
       "\n",
       "0                                 ENSG00000198695  ENSG00000198727  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                5               90   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                1               36   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                4               80   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                1               21   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0               15   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                6               97   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9               13               60   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                2               68   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0               27   \n",
       "\n",
       "0                                 ENSG00000215750  ENSG00000215689  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                1                0   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                0                0   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0                0   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0                0   \n",
       "\n",
       "0                                 ENSG00000215781  ENSG00000220023  \\\n",
       "0                                                                    \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                0                1   \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                1   \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                0                1   \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                2                0   \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0                1   \n",
       "...                                           ...              ...   \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                1                1   \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                1                1   \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                0                1   \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0   \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0                1   \n",
       "\n",
       "0                                 ENSG00000215615  ENSG00000215699  \n",
       "0                                                                   \n",
       "AAACCTGAGAGGTACC-1_1A_1_SI-GA-E2                0                0  \n",
       "AAACCTGAGTCAATAG-1_1A_1_SI-GA-E2                0                0  \n",
       "AAACCTGCAAACAACA-1_1A_1_SI-GA-E2                1                0  \n",
       "AAACCTGCACTTCTGC-1_1A_1_SI-GA-E2                0                0  \n",
       "AAACCTGCATGTAGTC-1_1A_1_SI-GA-E2                0                0  \n",
       "...                                           ...              ...  \n",
       "TTTGTCAGTACCTACA-1_2B_8_SI-GA-H9                0                0  \n",
       "TTTGTCAGTATCACCA-1_2B_8_SI-GA-H9                0                1  \n",
       "TTTGTCAGTTCAGACT-1_2B_8_SI-GA-H9                0                0  \n",
       "TTTGTCAGTTCTGTTT-1_2B_8_SI-GA-H9                0                0  \n",
       "TTTGTCATCAAAGTAG-1_2B_8_SI-GA-H9                0                0  \n",
       "\n",
       "[207324 rows x 13135 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exprs = pd.DataFrame.sparse.from_spmatrix(coo.T, index=cells, columns=genes)\n",
    "\n",
    "exprs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb048bd0-2b5d-42d0-b99f-fe64582abc54",
   "metadata": {},
   "source": [
    "### choose bassik_mch, a non-targeting control gRNA group and 100 genes as an example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5bc932d8-bdc8-4337-95ac-a01b2fd2c343",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenoData['gRNA_group'] = phenoData['gRNA_groups'].str.contains('bassik_mch', na=False).astype(int)\n",
    "\n",
    "subset = exprs[exprs.columns[:100]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce68b88a-6245-4ebe-8385-635b10f903c2",
   "metadata": {},
   "source": [
    "### differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a7ee575b-c084-4a2b-8a7f-7a2c39503fbf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [02:16<00:00,  1.36s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "      <th>term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0.121739</td>\n",
       "      <td>0.291886</td>\n",
       "      <td>0.417079</td>\n",
       "      <td>6.766204e-01</td>\n",
       "      <td>-0.450346</td>\n",
       "      <td>0.693825</td>\n",
       "      <td>gRNA_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>-0.000337</td>\n",
       "      <td>0.001084</td>\n",
       "      <td>-0.310907</td>\n",
       "      <td>7.558716e-01</td>\n",
       "      <td>-0.002462</td>\n",
       "      <td>0.001788</td>\n",
       "      <td>guide_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>11.765150</td>\n",
       "      <td>0.905019</td>\n",
       "      <td>12.999893</td>\n",
       "      <td>1.225147e-38</td>\n",
       "      <td>9.991345</td>\n",
       "      <td>13.538954</td>\n",
       "      <td>percent.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0.041200</td>\n",
       "      <td>0.032899</td>\n",
       "      <td>1.252321</td>\n",
       "      <td>2.104528e-01</td>\n",
       "      <td>-0.023281</td>\n",
       "      <td>0.105681</td>\n",
       "      <td>prep_batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>-4.495644</td>\n",
       "      <td>0.053886</td>\n",
       "      <td>-83.428977</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-4.601258</td>\n",
       "      <td>-4.390030</td>\n",
       "      <td>intercept</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-0.111505</td>\n",
       "      <td>0.093053</td>\n",
       "      <td>-1.198287</td>\n",
       "      <td>2.308052e-01</td>\n",
       "      <td>-0.293886</td>\n",
       "      <td>0.070877</td>\n",
       "      <td>gRNA_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-0.000184</td>\n",
       "      <td>0.000310</td>\n",
       "      <td>-0.593772</td>\n",
       "      <td>5.526647e-01</td>\n",
       "      <td>-0.000793</td>\n",
       "      <td>0.000424</td>\n",
       "      <td>guide_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>5.626938</td>\n",
       "      <td>0.263224</td>\n",
       "      <td>21.376973</td>\n",
       "      <td>2.188725e-101</td>\n",
       "      <td>5.111028</td>\n",
       "      <td>6.142848</td>\n",
       "      <td>percent.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>0.109529</td>\n",
       "      <td>0.009431</td>\n",
       "      <td>11.614294</td>\n",
       "      <td>3.486571e-31</td>\n",
       "      <td>0.091046</td>\n",
       "      <td>0.128013</td>\n",
       "      <td>prep_batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-1.548903</td>\n",
       "      <td>0.015096</td>\n",
       "      <td>-102.605106</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-1.578490</td>\n",
       "      <td>-1.519316</td>\n",
       "      <td>intercept</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Coef.  Std.Err.           z          P>|z|    [0.025  \\\n",
       "ENSG00000238009   0.121739  0.291886    0.417079   6.766204e-01 -0.450346   \n",
       "ENSG00000238009  -0.000337  0.001084   -0.310907   7.558716e-01 -0.002462   \n",
       "ENSG00000238009  11.765150  0.905019   12.999893   1.225147e-38  9.991345   \n",
       "ENSG00000238009   0.041200  0.032899    1.252321   2.104528e-01 -0.023281   \n",
       "ENSG00000238009  -4.495644  0.053886  -83.428977   0.000000e+00 -4.601258   \n",
       "...                    ...       ...         ...            ...       ...   \n",
       "ENSG00000142599  -0.111505  0.093053   -1.198287   2.308052e-01 -0.293886   \n",
       "ENSG00000142599  -0.000184  0.000310   -0.593772   5.526647e-01 -0.000793   \n",
       "ENSG00000142599   5.626938  0.263224   21.376973  2.188725e-101  5.111028   \n",
       "ENSG00000142599   0.109529  0.009431   11.614294   3.486571e-31  0.091046   \n",
       "ENSG00000142599  -1.548903  0.015096 -102.605106   0.000000e+00 -1.578490   \n",
       "\n",
       "                    0.975]          term  \n",
       "ENSG00000238009   0.693825    gRNA_group  \n",
       "ENSG00000238009   0.001788   guide_count  \n",
       "ENSG00000238009  13.538954  percent.mito  \n",
       "ENSG00000238009   0.105681    prep_batch  \n",
       "ENSG00000238009  -4.390030     intercept  \n",
       "...                    ...           ...  \n",
       "ENSG00000142599   0.070877    gRNA_group  \n",
       "ENSG00000142599   0.000424   guide_count  \n",
       "ENSG00000142599   6.142848  percent.mito  \n",
       "ENSG00000142599   0.128013    prep_batch  \n",
       "ENSG00000142599  -1.519316     intercept  \n",
       "\n",
       "[500 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# x\n",
    "\n",
    "x = phenoData[['gRNA_group','guide_count','percent.mito','prep_batch']].values\n",
    "x = sm.api.add_constant(x, prepend=False)\n",
    "\n",
    "xname = ['gRNA_group','guide_count','percent.mito','prep_batch','intercept']\n",
    "\n",
    "# y\n",
    "\n",
    "size_factor = phenoData['size_factor'].values\n",
    "\n",
    "Y = np.round(subset.values.T / size_factor)\n",
    "\n",
    "# fit\n",
    "\n",
    "summaries = list()\n",
    "for gene, y in zip(tqdm(subset.columns), Y):\n",
    "    \n",
    "    model = sm.discrete.discrete_model.NegativeBinomial(y, x, loglike_method='geometric')\n",
    "    res = model.fit(disp=False, warn_convergence=False)\n",
    "    summary = res.summary2(xname=xname).tables[1]\n",
    "    \n",
    "    summary['term'] = summary.index\n",
    "    summary.index = [gene] * summary.shape[0]\n",
    "    summaries.append(summary)\n",
    "    \n",
    "summaries = pd.concat(summaries, axis=0)\n",
    "\n",
    "summaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c0055f1-e69f-42b2-8fcf-156dff61f9df",
   "metadata": {},
   "source": [
    "### compare with monocle3 results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e00683d5-bfa9-4c38-88bd-7d5984f94a63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spearman correlation of\n",
      "effect size 0.9972397239723971 1.7145424785108527e-112\n",
      "intercept 0.9998799879987997 3.439317197857124e-179\n",
      "p-value 0.9475907590759075 2.314617210933284e-50\n"
     ]
    }
   ],
   "source": [
    "# load monocle3 results\n",
    "df = pd.read_csv('bassik_mch-100.csv', index_col=0)\n",
    "\n",
    "# transform into wide-style table\n",
    "\n",
    "monocle3_results = df.query('term == \"gRNA_group\"')[['term','estimate','p_value']]\n",
    "for i in ['(Intercept)','guide_count','percent.mito','prep_batchprep_batch_2']:\n",
    "    monocle3_results[i] = df.query('term == \"%s\"' % i)['estimate']\n",
    "    \n",
    "results = summaries.query('term == \"gRNA_group\"')[['term','Coef.','P>|z|']]\n",
    "for i in ['intercept','guide_count','percent.mito','prep_batch']:\n",
    "    results[i] = summaries.query('term == \"%s\"' % i)['Coef.']\n",
    "    \n",
    "# Pearson correlation test\n",
    "\n",
    "print('Spearman correlation of')\n",
    "print('effect size', *stats.spearmanr(results['Coef.'], monocle3_results['estimate']))\n",
    "print('intercept', *stats.spearmanr(results['intercept'], monocle3_results['(Intercept)']))\n",
    "print('p-value', *stats.spearmanr(results['P>|z|'], monocle3_results['p_value']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8465080-00a5-45fe-a54f-f7a66a26e457",
   "metadata": {},
   "source": [
    "Statsmodels got highly consistent results with monocle3, but it was still too slow.\n",
    "\n",
    "Now, let's try scpy."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4463a3a-7015-41f5-8af0-95e114cacb08",
   "metadata": {},
   "source": [
    "## Scpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f35f63bb-d125-4459-8156-05ee8cd57d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import scpy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87370285-2bc6-4374-8ce0-d80ec7a94551",
   "metadata": {},
   "source": [
    "### differential expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b640ca9f-7aa1-436c-bcdd-5abbf85dd849",
   "metadata": {},
   "source": [
    "Scpy uses an API similar to statsmodels when fitting a single pair."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "aad32baa-25f3-410b-a6ce-4f551574a34c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [00:10<00:00,  9.10it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0.118145</td>\n",
       "      <td>0.292382</td>\n",
       "      <td>0.404078</td>\n",
       "      <td>6.861557e-01</td>\n",
       "      <td>gRNA_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>-0.000333</td>\n",
       "      <td>0.001084</td>\n",
       "      <td>-0.306772</td>\n",
       "      <td>7.590171e-01</td>\n",
       "      <td>guide_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>11.762953</td>\n",
       "      <td>0.905033</td>\n",
       "      <td>12.997268</td>\n",
       "      <td>1.312820e-38</td>\n",
       "      <td>percent.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0.041310</td>\n",
       "      <td>0.032900</td>\n",
       "      <td>1.255632</td>\n",
       "      <td>2.092508e-01</td>\n",
       "      <td>prep_batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>-4.495757</td>\n",
       "      <td>0.053887</td>\n",
       "      <td>-83.429970</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>intercept</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-0.113176</td>\n",
       "      <td>0.093100</td>\n",
       "      <td>-1.215645</td>\n",
       "      <td>2.241214e-01</td>\n",
       "      <td>gRNA_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-0.000190</td>\n",
       "      <td>0.000311</td>\n",
       "      <td>-0.611838</td>\n",
       "      <td>5.406457e-01</td>\n",
       "      <td>guide_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>5.665720</td>\n",
       "      <td>0.263223</td>\n",
       "      <td>21.524412</td>\n",
       "      <td>1.192395e-102</td>\n",
       "      <td>percent.mito</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>0.109481</td>\n",
       "      <td>0.009431</td>\n",
       "      <td>11.609066</td>\n",
       "      <td>3.789701e-31</td>\n",
       "      <td>prep_batch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-1.550258</td>\n",
       "      <td>0.015097</td>\n",
       "      <td>-102.684689</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>intercept</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Coef.  Std.Err.           z          P>|z|          term\n",
       "ENSG00000238009   0.118145  0.292382    0.404078   6.861557e-01    gRNA_group\n",
       "ENSG00000238009  -0.000333  0.001084   -0.306772   7.590171e-01   guide_count\n",
       "ENSG00000238009  11.762953  0.905033   12.997268   1.312820e-38  percent.mito\n",
       "ENSG00000238009   0.041310  0.032900    1.255632   2.092508e-01    prep_batch\n",
       "ENSG00000238009  -4.495757  0.053887  -83.429970   0.000000e+00     intercept\n",
       "...                    ...       ...         ...            ...           ...\n",
       "ENSG00000142599  -0.113176  0.093100   -1.215645   2.241214e-01    gRNA_group\n",
       "ENSG00000142599  -0.000190  0.000311   -0.611838   5.406457e-01   guide_count\n",
       "ENSG00000142599   5.665720  0.263223   21.524412  1.192395e-102  percent.mito\n",
       "ENSG00000142599   0.109481  0.009431   11.609066   3.789701e-31    prep_batch\n",
       "ENSG00000142599  -1.550258  0.015097 -102.684689   0.000000e+00     intercept\n",
       "\n",
       "[500 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summaries = list()\n",
    "for gene, y in zip(tqdm(subset.columns), Y):\n",
    "    \n",
    "    summary = scpy.fit(x, y, 'negative_binomial', device='cpu', xname=xname)\n",
    "    \n",
    "    summary['term'] = summary.index\n",
    "    summary.index = [gene] * summary.shape[0]\n",
    "    summaries.append(summary)\n",
    "    \n",
    "summaries = pd.concat(summaries, axis=0)\n",
    "\n",
    "summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "08ff6c0a-b5b9-4906-9f37-7fbbc64b1e73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spearman correlation of\n",
      "effect size 0.9985718571857185 1.678441941676762e-126\n",
      "intercept 0.9998919891989198 1.9700524468233987e-181\n",
      "p-value 0.9574557455745574 1.071068835741778e-54\n"
     ]
    }
   ],
   "source": [
    "results = summaries.query('term == \"gRNA_group\"')[['term','Coef.','P>|z|']]\n",
    "for i in ['intercept','guide_count','percent.mito','prep_batch']:\n",
    "    results[i] = summaries.query('term == \"%s\"' % i)['Coef.']\n",
    "    \n",
    "# Pearson correlation test\n",
    "\n",
    "print('Spearman correlation of')\n",
    "print('effect size', *stats.spearmanr(results['Coef.'], monocle3_results['estimate']))\n",
    "print('intercept', *stats.spearmanr(results['intercept'], monocle3_results['(Intercept)']))\n",
    "print('p-value', *stats.spearmanr(results['P>|z|'], monocle3_results['p_value']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d667364-3e79-496c-ae5d-7bd2827814ce",
   "metadata": {},
   "source": [
    "Wow, it's **20X faster** and as accurate as monocle3."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d896cece-b4f0-470f-ab12-8441b48965d8",
   "metadata": {},
   "source": [
    "### differential expression analysis in batch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e89d1941-6ed0-4790-848e-70303b5a7049",
   "metadata": {},
   "source": [
    "Scpy provides a more convenient and efficient API for fitting multiple sets of independent variables and dependent variables at the same time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ab1b388b-e75e-43e2-82b6-d13610f0fc20",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [00:05<00:00, 16.95it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group</th>\n",
       "      <th>term</th>\n",
       "      <th>target</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">bassik_mch</th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0.118145</td>\n",
       "      <td>0.292382</td>\n",
       "      <td>0.404078</td>\n",
       "      <td>6.861557e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>guide_count</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>-0.000333</td>\n",
       "      <td>0.001084</td>\n",
       "      <td>-0.306772</td>\n",
       "      <td>7.590171e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>percent.mito</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>11.762953</td>\n",
       "      <td>0.905033</td>\n",
       "      <td>12.997268</td>\n",
       "      <td>1.312820e-38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prep_batch</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>0.041310</td>\n",
       "      <td>0.032900</td>\n",
       "      <td>1.255632</td>\n",
       "      <td>2.092508e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intercept</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>-4.495757</td>\n",
       "      <td>0.053887</td>\n",
       "      <td>-83.429970</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-0.113176</td>\n",
       "      <td>0.093100</td>\n",
       "      <td>-1.215645</td>\n",
       "      <td>2.241214e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>guide_count</th>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-0.000190</td>\n",
       "      <td>0.000311</td>\n",
       "      <td>-0.611838</td>\n",
       "      <td>5.406457e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>percent.mito</th>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>5.665720</td>\n",
       "      <td>0.263223</td>\n",
       "      <td>21.524412</td>\n",
       "      <td>1.192395e-102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prep_batch</th>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>0.109481</td>\n",
       "      <td>0.009431</td>\n",
       "      <td>11.609066</td>\n",
       "      <td>3.789701e-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intercept</th>\n",
       "      <th>ENSG00000142599</th>\n",
       "      <td>-1.550258</td>\n",
       "      <td>0.015097</td>\n",
       "      <td>-102.684689</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Coef.  Std.Err.           z  \\\n",
       "group      term         target                                             \n",
       "bassik_mch gRNA_group   ENSG00000238009   0.118145  0.292382    0.404078   \n",
       "           guide_count  ENSG00000238009  -0.000333  0.001084   -0.306772   \n",
       "           percent.mito ENSG00000238009  11.762953  0.905033   12.997268   \n",
       "           prep_batch   ENSG00000238009   0.041310  0.032900    1.255632   \n",
       "           intercept    ENSG00000238009  -4.495757  0.053887  -83.429970   \n",
       "...                                            ...       ...         ...   \n",
       "           gRNA_group   ENSG00000142599  -0.113176  0.093100   -1.215645   \n",
       "           guide_count  ENSG00000142599  -0.000190  0.000311   -0.611838   \n",
       "           percent.mito ENSG00000142599   5.665720  0.263223   21.524412   \n",
       "           prep_batch   ENSG00000142599   0.109481  0.009431   11.609066   \n",
       "           intercept    ENSG00000142599  -1.550258  0.015097 -102.684689   \n",
       "\n",
       "                                                 P>|z|  \n",
       "group      term         target                          \n",
       "bassik_mch gRNA_group   ENSG00000238009   6.861557e-01  \n",
       "           guide_count  ENSG00000238009   7.590171e-01  \n",
       "           percent.mito ENSG00000238009   1.312820e-38  \n",
       "           prep_batch   ENSG00000238009   2.092508e-01  \n",
       "           intercept    ENSG00000238009   0.000000e+00  \n",
       "...                                                ...  \n",
       "           gRNA_group   ENSG00000142599   2.241214e-01  \n",
       "           guide_count  ENSG00000142599   5.406457e-01  \n",
       "           percent.mito ENSG00000142599  1.192395e-102  \n",
       "           prep_batch   ENSG00000142599   3.789701e-31  \n",
       "           intercept    ENSG00000142599   0.000000e+00  \n",
       "\n",
       "[500 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summaries = scpy.fit_batch([x], Y, 'negative_binomial', device='cpu', xnames=[('bassik_mch', xname)], yname=subset.columns)\n",
    "\n",
    "summaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0581abc3-0078-41c2-b3cb-4da8dbf90213",
   "metadata": {},
   "source": [
    "If you have a GPU, scpy can run faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d5ea8cee-3308-476f-93d7-840715818ccb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [00:01<00:00, 54.09it/s]\n"
     ]
    }
   ],
   "source": [
    "summaries = scpy.fit_batch([x], Y, 'negative_binomial', device=0, xnames=[('bassik_mch', xname)], yname=subset.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cab93d5-0056-4ae4-a195-247b8fe9606c",
   "metadata": {},
   "source": [
    "Sometimes, the size of the dependent variable is large, or it takes a long time for preprocessing.\n",
    "\n",
    "For example, the expression matrix in the sample occupies 6.5 GB of memory when stored in a sparse form. If it is converted to dense form all at once for calculation, it will occupy at least 30GB of memory.\n",
    "\n",
    "Scpy provides lazy data loader to solve this problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5cfa2946-20e7-48c8-8ef9-69f971a8f629",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 100/100 [00:02<00:00, 47.75it/s]\n"
     ]
    }
   ],
   "source": [
    "loader = lambda i: np.round(subset[subset.columns[i]].values.T / size_factor)\n",
    "Ys = scpy.dataloader(loader, len(subset.columns), num_workers=2)\n",
    "\n",
    "summaries = scpy.fit_batch([x], Ys, 'negative_binomial', device=0, xnames=[('bassik_mch', xname)], yname=subset.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1ed6330-b60a-41d2-a102-7caef936ebcd",
   "metadata": {},
   "source": [
    "So far, we have been able to obtain **100X faster** than the R package.\n",
    "\n",
    "Can it be faster?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00ebc9ad-f5c7-4dda-903d-256216a2900c",
   "metadata": {},
   "source": [
    "### differential expression analysis in parallel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98e1984d-e2b4-4132-bcc3-ddc64a5ce191",
   "metadata": {},
   "source": [
    "In order to maximize the performance of scpy and the device, you can use the parallel API.\n",
    "\n",
    "Parallel scpy only provides a command line API based on multiprocessing, and requires a python script to export args.\n",
    "\n",
    "We provide a sample script, *demo.py* ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7169b6e5-6a12-4a24-bcd4-962435abda25",
   "metadata": {},
   "source": [
    "```python\n",
    "# prepare\n",
    "\n",
    "## x\n",
    "\n",
    "# select 2 gRNA groups targeting non-coding regions, 2 positiv control gRNA groups and 3 non-targeting control gRNA\n",
    "gRNA_groups = ['chr1.8492','chrX.952','CERS2_TSS','GATA2_TSS','bassik_mch','random_1','scrambled_1']\n",
    "\n",
    "X, xnames = list(), list()\n",
    "\n",
    "for gRNA_group in gRNA_groups:\n",
    "  \n",
    "  geno = phenoData['gRNA_groups'].str.contains(gRNA_group, na=False) # get genotypes\n",
    "  \n",
    "  x = geno.astype(int).values.reshape((-1, 1)) # use genotype as x\n",
    "  cov = phenoData[['guide_count','percent.mito','prep_batch']].values # covariants\n",
    "  c = np.ones((x.shape[0], 1)) # constraint\n",
    "\n",
    "  X.append(np.concatenate([x, cov, c], axis=1))\n",
    "  xnames.append((gRNA_group, ['gRNA_group', 'guide_count', 'percent.mito', 'prep_batch', 'intercept']))\n",
    "  \n",
    "## y\n",
    "\n",
    "size_factor = phenoData['Size_Factor'].values # load size factor calculate by monocle\n",
    "\n",
    "loader = lambda i: np.round(exprs[exprs.columns[i]].values.T / size_factor)\n",
    "\n",
    "\n",
    "# export\n",
    "\n",
    "args = {\n",
    "  'X': X,\n",
    "  'Y': {'data': loader, 'length': exprs.shape[1]},\n",
    "  'family': 'negative_binomial',\n",
    "  'xnames': xnames,\n",
    "  'yname': exprs.columns\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ffe4d0b-6d3e-40db-9912-be434d3d81bb",
   "metadata": {},
   "source": [
    "If I have 8 GPUs, I can run the following commands in the console:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b12a20b-b9b6-40f8-bd6a-56939f911d69",
   "metadata": {},
   "source": [
    "```bash\n",
    "python -m scpy demo.py demo.csv -d 0,1,2,3,4,5,6,7 --disp=True --low_memory=True\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcfd6381-d593-421f-b043-bc9cf4453e53",
   "metadata": {},
   "source": [
    "```\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1642/1642 [03:31<00:00,  7.77it/s]\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1642/1642 [03:31<00:00,  7.77it/s]\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1642/1642 [03:32<00:00,  7.72it/s]\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1642/1642 [03:33<00:00,  7.69it/s]\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1642/1642 [03:32<00:00,  7.71it/s]\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1641/1641 [03:33<00:00,  7.68it/s]\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1642/1642 [03:33<00:00,  7.69it/s]\n",
    "100%|█████████████████████████████████████████████████████████████████████████████████| 1642/1642 [03:34<00:00,  7.66it/s]\n",
    "221.72746539115906\n",
    "                                                    Coef.  Std.Err.          z         P>|z|\n",
    "group             term         target\n",
    "chr1.8492_top_two gRNA_group   ENSG00000238009   0.202458  0.196002   1.032938  3.016340e-01\n",
    "                  guide_count  ENSG00000238009   0.008593  0.001002   8.579515  9.591534e-18\n",
    "                  percent.mito ENSG00000238009  17.562862  0.873383  20.109011  7.500587e-90\n",
    "                  prep_batch   ENSG00000238009  -0.019360  0.031808  -0.608657  5.427524e-01\n",
    "                  intercept    ENSG00000238009  -4.929867  0.053910 -91.446486  0.000000e+00\n",
    "...                                                   ...       ...        ...           ...\n",
    "bassik_mch        gRNA_group   ENSG00000215699   0.388362  0.320037   1.213490  2.249438e-01\n",
    "                  guide_count  ENSG00000215699   0.006148  0.001344   4.574964  4.765763e-06\n",
    "                  percent.mito ENSG00000215699  13.970382  1.161668  12.026145  2.656786e-33\n",
    "                  prep_batch   ENSG00000215699   0.043774  0.042161   1.038266  2.991475e-01\n",
    "                  intercept    ENSG00000215699  -5.298037  0.070448 -75.204969  0.000000e+00\n",
    "\n",
    "[459725 rows x 4 columns]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9f2d631-075f-4ea4-b962-96e71a054b2c",
   "metadata": {},
   "source": [
    "**OMG, we are now 1000X faster ! ! !**\n",
    "\n",
    "Let us check whether the results of scpy are consistent with those provided by Gasperini et al."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3e7a2144-6cb8-4a6d-ac0f-f63fa498d6c5",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jjpeng/data/.conda/envs/torch-latest/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (12,13) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>beta</th>\n",
       "      <th>intercept</th>\n",
       "      <th>fold_change.transcript_remaining</th>\n",
       "      <th>pvalue.raw</th>\n",
       "      <th>pvalue.empirical</th>\n",
       "      <th>pvalue.empirical.adjusted</th>\n",
       "      <th>ENSG</th>\n",
       "      <th>gene_short_name</th>\n",
       "      <th>gRNA_group</th>\n",
       "      <th>quality_rank_grna</th>\n",
       "      <th>target_site.chr</th>\n",
       "      <th>target_site.start</th>\n",
       "      <th>target_site.stop</th>\n",
       "      <th>target_gene.chr</th>\n",
       "      <th>target_gene.start</th>\n",
       "      <th>target_gene.stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>site_type</th>\n",
       "      <th>outlier_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pairs4merge</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACTB_TSS:ENSG00000008256</th>\n",
       "      <td>-0.114784</td>\n",
       "      <td>-3.432141</td>\n",
       "      <td>0.891559</td>\n",
       "      <td>2.746239e-01</td>\n",
       "      <td>0.261969034</td>\n",
       "      <td>0.831159512731444</td>\n",
       "      <td>ENSG00000008256</td>\n",
       "      <td>CYTH3</td>\n",
       "      <td>ACTB_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5570339</td>\n",
       "      <td>5570340</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6312274</td>\n",
       "      <td>6312275</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB_TSS:ENSG00000011275</th>\n",
       "      <td>-0.042450</td>\n",
       "      <td>-1.464958</td>\n",
       "      <td>0.958438</td>\n",
       "      <td>2.771529e-01</td>\n",
       "      <td>0.264361067</td>\n",
       "      <td>0.831814624048016</td>\n",
       "      <td>ENSG00000011275</td>\n",
       "      <td>RNF216</td>\n",
       "      <td>ACTB_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5570339</td>\n",
       "      <td>5570340</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5821369</td>\n",
       "      <td>5821370</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB_TSS:ENSG00000075618</th>\n",
       "      <td>0.034076</td>\n",
       "      <td>0.392365</td>\n",
       "      <td>1.034664</td>\n",
       "      <td>1.425945e-01</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>ENSG00000075618</td>\n",
       "      <td>FSCN1</td>\n",
       "      <td>ACTB_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5570339</td>\n",
       "      <td>5570340</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5632454</td>\n",
       "      <td>5632455</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB_TSS:ENSG00000075624</th>\n",
       "      <td>-0.437939</td>\n",
       "      <td>2.898258</td>\n",
       "      <td>0.645365</td>\n",
       "      <td>7.463645e-115</td>\n",
       "      <td>3.941e-06</td>\n",
       "      <td>0.000407265538461538</td>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>ACTB_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5570339</td>\n",
       "      <td>5570340</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5570339</td>\n",
       "      <td>5570340</td>\n",
       "      <td>-</td>\n",
       "      <td>selfTSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB_TSS:ENSG00000086232</th>\n",
       "      <td>-0.010209</td>\n",
       "      <td>1.154783</td>\n",
       "      <td>0.989843</td>\n",
       "      <td>6.253678e-01</td>\n",
       "      <td>0.611785198</td>\n",
       "      <td>0.947813794532952</td>\n",
       "      <td>ENSG00000086232</td>\n",
       "      <td>EIF2AK1</td>\n",
       "      <td>ACTB_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5570339</td>\n",
       "      <td>5570340</td>\n",
       "      <td>chr7</td>\n",
       "      <td>6098860</td>\n",
       "      <td>6098861</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB2_TSS:ENSG00000116754</th>\n",
       "      <td>-0.016496</td>\n",
       "      <td>0.174595</td>\n",
       "      <td>0.983639</td>\n",
       "      <td>4.979588e-01</td>\n",
       "      <td>0.482804551</td>\n",
       "      <td>0.916882597262894</td>\n",
       "      <td>ENSG00000116754</td>\n",
       "      <td>SRSF11</td>\n",
       "      <td>ZRANB2_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr1</td>\n",
       "      <td>71546971</td>\n",
       "      <td>71546972</td>\n",
       "      <td>chr1</td>\n",
       "      <td>70671365</td>\n",
       "      <td>70671366</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB2_TSS:ENSG00000116761</th>\n",
       "      <td>0.022505</td>\n",
       "      <td>-1.795345</td>\n",
       "      <td>1.022760</td>\n",
       "      <td>7.781657e-01</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>ENSG00000116761</td>\n",
       "      <td>CTH</td>\n",
       "      <td>ZRANB2_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr1</td>\n",
       "      <td>71546971</td>\n",
       "      <td>71546972</td>\n",
       "      <td>chr1</td>\n",
       "      <td>70876901</td>\n",
       "      <td>70876902</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB2_TSS:ENSG00000118454</th>\n",
       "      <td>-0.039160</td>\n",
       "      <td>-2.665878</td>\n",
       "      <td>0.961597</td>\n",
       "      <td>6.191715e-01</td>\n",
       "      <td>0.605440595</td>\n",
       "      <td>0.946588425199907</td>\n",
       "      <td>ENSG00000118454</td>\n",
       "      <td>ANKRD13C</td>\n",
       "      <td>ZRANB2_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr1</td>\n",
       "      <td>71546971</td>\n",
       "      <td>71546972</td>\n",
       "      <td>chr1</td>\n",
       "      <td>70820416</td>\n",
       "      <td>70820417</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB2_TSS:ENSG00000132485</th>\n",
       "      <td>-0.581223</td>\n",
       "      <td>-0.493145</td>\n",
       "      <td>0.559214</td>\n",
       "      <td>2.525892e-65</td>\n",
       "      <td>3.941e-06</td>\n",
       "      <td>0.000407265538461538</td>\n",
       "      <td>ENSG00000132485</td>\n",
       "      <td>ZRANB2</td>\n",
       "      <td>ZRANB2_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr1</td>\n",
       "      <td>71546971</td>\n",
       "      <td>71546972</td>\n",
       "      <td>chr1</td>\n",
       "      <td>71546971</td>\n",
       "      <td>71546972</td>\n",
       "      <td>-</td>\n",
       "      <td>selfTSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB2_TSS:ENSG00000197568</th>\n",
       "      <td>0.052511</td>\n",
       "      <td>-1.413776</td>\n",
       "      <td>1.053914</td>\n",
       "      <td>3.484490e-01</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>ENSG00000197568</td>\n",
       "      <td>HHLA3</td>\n",
       "      <td>ZRANB2_TSS</td>\n",
       "      <td>control</td>\n",
       "      <td>chr1</td>\n",
       "      <td>71546971</td>\n",
       "      <td>71546972</td>\n",
       "      <td>chr1</td>\n",
       "      <td>70820488</td>\n",
       "      <td>70820489</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>629515 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                beta  intercept  \\\n",
       "pairs4merge                                       \n",
       "ACTB_TSS:ENSG00000008256   -0.114784  -3.432141   \n",
       "ACTB_TSS:ENSG00000011275   -0.042450  -1.464958   \n",
       "ACTB_TSS:ENSG00000075618    0.034076   0.392365   \n",
       "ACTB_TSS:ENSG00000075624   -0.437939   2.898258   \n",
       "ACTB_TSS:ENSG00000086232   -0.010209   1.154783   \n",
       "...                              ...        ...   \n",
       "ZRANB2_TSS:ENSG00000116754 -0.016496   0.174595   \n",
       "ZRANB2_TSS:ENSG00000116761  0.022505  -1.795345   \n",
       "ZRANB2_TSS:ENSG00000118454 -0.039160  -2.665878   \n",
       "ZRANB2_TSS:ENSG00000132485 -0.581223  -0.493145   \n",
       "ZRANB2_TSS:ENSG00000197568  0.052511  -1.413776   \n",
       "\n",
       "                            fold_change.transcript_remaining     pvalue.raw  \\\n",
       "pairs4merge                                                                   \n",
       "ACTB_TSS:ENSG00000008256                            0.891559   2.746239e-01   \n",
       "ACTB_TSS:ENSG00000011275                            0.958438   2.771529e-01   \n",
       "ACTB_TSS:ENSG00000075618                            1.034664   1.425945e-01   \n",
       "ACTB_TSS:ENSG00000075624                            0.645365  7.463645e-115   \n",
       "ACTB_TSS:ENSG00000086232                            0.989843   6.253678e-01   \n",
       "...                                                      ...            ...   \n",
       "ZRANB2_TSS:ENSG00000116754                          0.983639   4.979588e-01   \n",
       "ZRANB2_TSS:ENSG00000116761                          1.022760   7.781657e-01   \n",
       "ZRANB2_TSS:ENSG00000118454                          0.961597   6.191715e-01   \n",
       "ZRANB2_TSS:ENSG00000132485                          0.559214   2.525892e-65   \n",
       "ZRANB2_TSS:ENSG00000197568                          1.053914   3.484490e-01   \n",
       "\n",
       "                           pvalue.empirical pvalue.empirical.adjusted  \\\n",
       "pairs4merge                                                             \n",
       "ACTB_TSS:ENSG00000008256        0.261969034         0.831159512731444   \n",
       "ACTB_TSS:ENSG00000011275        0.264361067         0.831814624048016   \n",
       "ACTB_TSS:ENSG00000075618     not_applicable            not_applicable   \n",
       "ACTB_TSS:ENSG00000075624          3.941e-06      0.000407265538461538   \n",
       "ACTB_TSS:ENSG00000086232        0.611785198         0.947813794532952   \n",
       "...                                     ...                       ...   \n",
       "ZRANB2_TSS:ENSG00000116754      0.482804551         0.916882597262894   \n",
       "ZRANB2_TSS:ENSG00000116761   not_applicable            not_applicable   \n",
       "ZRANB2_TSS:ENSG00000118454      0.605440595         0.946588425199907   \n",
       "ZRANB2_TSS:ENSG00000132485        3.941e-06      0.000407265538461538   \n",
       "ZRANB2_TSS:ENSG00000197568   not_applicable            not_applicable   \n",
       "\n",
       "                                       ENSG gene_short_name  gRNA_group  \\\n",
       "pairs4merge                                                               \n",
       "ACTB_TSS:ENSG00000008256    ENSG00000008256           CYTH3    ACTB_TSS   \n",
       "ACTB_TSS:ENSG00000011275    ENSG00000011275          RNF216    ACTB_TSS   \n",
       "ACTB_TSS:ENSG00000075618    ENSG00000075618           FSCN1    ACTB_TSS   \n",
       "ACTB_TSS:ENSG00000075624    ENSG00000075624            ACTB    ACTB_TSS   \n",
       "ACTB_TSS:ENSG00000086232    ENSG00000086232         EIF2AK1    ACTB_TSS   \n",
       "...                                     ...             ...         ...   \n",
       "ZRANB2_TSS:ENSG00000116754  ENSG00000116754          SRSF11  ZRANB2_TSS   \n",
       "ZRANB2_TSS:ENSG00000116761  ENSG00000116761             CTH  ZRANB2_TSS   \n",
       "ZRANB2_TSS:ENSG00000118454  ENSG00000118454        ANKRD13C  ZRANB2_TSS   \n",
       "ZRANB2_TSS:ENSG00000132485  ENSG00000132485          ZRANB2  ZRANB2_TSS   \n",
       "ZRANB2_TSS:ENSG00000197568  ENSG00000197568           HHLA3  ZRANB2_TSS   \n",
       "\n",
       "                           quality_rank_grna target_site.chr  \\\n",
       "pairs4merge                                                    \n",
       "ACTB_TSS:ENSG00000008256             control            chr7   \n",
       "ACTB_TSS:ENSG00000011275             control            chr7   \n",
       "ACTB_TSS:ENSG00000075618             control            chr7   \n",
       "ACTB_TSS:ENSG00000075624             control            chr7   \n",
       "ACTB_TSS:ENSG00000086232             control            chr7   \n",
       "...                                      ...             ...   \n",
       "ZRANB2_TSS:ENSG00000116754           control            chr1   \n",
       "ZRANB2_TSS:ENSG00000116761           control            chr1   \n",
       "ZRANB2_TSS:ENSG00000118454           control            chr1   \n",
       "ZRANB2_TSS:ENSG00000132485           control            chr1   \n",
       "ZRANB2_TSS:ENSG00000197568           control            chr1   \n",
       "\n",
       "                           target_site.start target_site.stop target_gene.chr  \\\n",
       "pairs4merge                                                                     \n",
       "ACTB_TSS:ENSG00000008256             5570339          5570340            chr7   \n",
       "ACTB_TSS:ENSG00000011275             5570339          5570340            chr7   \n",
       "ACTB_TSS:ENSG00000075618             5570339          5570340            chr7   \n",
       "ACTB_TSS:ENSG00000075624             5570339          5570340            chr7   \n",
       "ACTB_TSS:ENSG00000086232             5570339          5570340            chr7   \n",
       "...                                      ...              ...             ...   \n",
       "ZRANB2_TSS:ENSG00000116754          71546971         71546972            chr1   \n",
       "ZRANB2_TSS:ENSG00000116761          71546971         71546972            chr1   \n",
       "ZRANB2_TSS:ENSG00000118454          71546971         71546972            chr1   \n",
       "ZRANB2_TSS:ENSG00000132485          71546971         71546972            chr1   \n",
       "ZRANB2_TSS:ENSG00000197568          71546971         71546972            chr1   \n",
       "\n",
       "                            target_gene.start  target_gene.stop strand  \\\n",
       "pairs4merge                                                              \n",
       "ACTB_TSS:ENSG00000008256              6312274           6312275      -   \n",
       "ACTB_TSS:ENSG00000011275              5821369           5821370      -   \n",
       "ACTB_TSS:ENSG00000075618              5632454           5632455      +   \n",
       "ACTB_TSS:ENSG00000075624              5570339           5570340      -   \n",
       "ACTB_TSS:ENSG00000086232              6098860           6098861      -   \n",
       "...                                       ...               ...    ...   \n",
       "ZRANB2_TSS:ENSG00000116754           70671365          70671366      +   \n",
       "ZRANB2_TSS:ENSG00000116761           70876901          70876902      +   \n",
       "ZRANB2_TSS:ENSG00000118454           70820416          70820417      -   \n",
       "ZRANB2_TSS:ENSG00000132485           71546971          71546972      -   \n",
       "ZRANB2_TSS:ENSG00000197568           70820488          70820489      +   \n",
       "\n",
       "                           site_type  outlier_gene  \n",
       "pairs4merge                                         \n",
       "ACTB_TSS:ENSG00000008256         TSS         False  \n",
       "ACTB_TSS:ENSG00000011275         TSS         False  \n",
       "ACTB_TSS:ENSG00000075618         TSS         False  \n",
       "ACTB_TSS:ENSG00000075624     selfTSS         False  \n",
       "ACTB_TSS:ENSG00000086232         TSS         False  \n",
       "...                              ...           ...  \n",
       "ZRANB2_TSS:ENSG00000116754       TSS         False  \n",
       "ZRANB2_TSS:ENSG00000116761       TSS         False  \n",
       "ZRANB2_TSS:ENSG00000118454       TSS         False  \n",
       "ZRANB2_TSS:ENSG00000132485   selfTSS         False  \n",
       "ZRANB2_TSS:ENSG00000197568       TSS         False  \n",
       "\n",
       "[629515 rows x 19 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load results from the paper\n",
    "\n",
    "paper = pd.read_csv(base + 'GSE120861_all_deg_results.at_scale.txt.gz', sep='\\t', index_col='pairs4merge')\n",
    "\n",
    "paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2aea8355-230f-4577-b0ba-e25ad41567e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " CERS2_TSS \n",
      "\n",
      "effect size 0.9823353050806256 9.800019237294788e-31\n",
      "intercept 0.998541447208492 2.4632354748867975e-52\n",
      "p-value 0.95753990762499 3.235472204989662e-23\n",
      "\n",
      " GATA2_TSS \n",
      "\n",
      "effect size 0.9964912280701754 9.285165080058482e-20\n",
      "intercept 0.9982456140350877 2.579846712328762e-22\n",
      "p-value 0.9210526315789472 2.2361332893791837e-08\n",
      "\n",
      " bassik_mch \n",
      "\n",
      "effect size 0.9879195444074778 0.0\n",
      "intercept 0.9994431151886197 0.0\n",
      "p-value 0.9590769458326537 0.0\n",
      "\n",
      " chr1.8492_second_two \n",
      "\n",
      "effect size 0.9931934203062961 5.631027405850918e-39\n",
      "intercept 0.9987035086297708 2.339336383494484e-53\n",
      "p-value 0.9530021878291874 2.3641247234831865e-22\n",
      "\n",
      " chr1.8492_top_two \n",
      "\n",
      "effect size 0.9701806984847258 3.095784714315404e-26\n",
      "intercept 0.9987035086297708 2.339336383494484e-53\n",
      "p-value 0.9701806984847258 3.095784714315404e-26\n",
      "\n",
      " chrX.952_second_two \n",
      "\n",
      "effect size 0.9952346041055717 6.809255742590661e-32\n",
      "intercept 0.9978005865102637 6.361252877369035e-37\n",
      "p-value 0.9754398826979471 2.8726637865941426e-21\n",
      "\n",
      " chrX.952_top_two \n",
      "\n",
      "effect size 0.9882697947214076 4.800382460030913e-26\n",
      "intercept 0.9978005865102637 6.361252877369035e-37\n",
      "p-value 0.9772727272727272 9.084840026751097e-22\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('demo.csv')\n",
    "\n",
    "for gRNA_group, g in df.groupby('group'):\n",
    "    \n",
    "    g.set_index('target', inplace=True)\n",
    "    \n",
    "    results = g.query('term == \"gRNA_group\"')[['term','Coef.','P>|z|']]\n",
    "    results['intercept'] = g.query('term == \"intercept\"')['Coef.']\n",
    "    for i in ['guide_count','percent.mito','prep_batch']:\n",
    "        results[i] = g.query('term == \"%s\"' % i)['Coef.']\n",
    "        \n",
    "    paper_results = paper.query('gRNA_group == \"%s\"' % gRNA_group)\n",
    "    paper_results.set_index('ENSG', inplace=True)\n",
    "    paper_results = paper_results.loc[paper_results.index.intersection(results.index)]\n",
    "    \n",
    "    results = results.loc[paper_results.index]\n",
    "    \n",
    "    print('\\n', gRNA_group, '\\n')\n",
    "\n",
    "    print('effect size', *stats.spearmanr(results['Coef.'], paper_results['beta']))\n",
    "    print('intercept', *stats.spearmanr(results['intercept'], paper_results['intercept']))\n",
    "    print('p-value', *stats.spearmanr(results['P>|z|'], paper_results['pvalue.raw']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83eb6ca7-7bc1-4504-8e9d-79b772cd666f",
   "metadata": {},
   "source": [
    "## Reference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f849794f-2938-4da8-97cb-2be7b977201c",
   "metadata": {},
   "source": [
    "[1] Gasperini M, Hill AJ, McFaline-Figueroa JL, Martin B et al. A Genome-wide Framework for Mapping Gene Regulation via Cellular Genetic Screens. Cell 2019 Jan 10;176(1-2):377-390.e19. PMID: 30612741\n",
    "\n",
    "[2] Qiu, X. et. al. Reversed graph embedding resolves complex single-cell trajectories. Nat. Methods 14, 979–982 (2017). https://doi.org/10.1038/nmeth.4402\n",
    "\n",
    "[3] Cao, J. et. al. The single-cell transcriptional landscape of mammalian organogenesis. Nature 566, 496–502 (2019). https://doi.org/10.1038/s41586-019-0969-x\n",
    "\n",
    "[4] Seabold, Skipper, and Josef Perktold. “statsmodels: Econometric and statistical modeling with python.” Proceedings of the 9th Python in Science Conference. 2010."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "torch-latest",
   "language": "python",
   "name": "torch-latest"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
